// @noction/vue-draggable-grid/dist/index.js
var ni = Object.defineProperty;
var oi = (e, t, n) => t in e ? ni(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n;
var d = (e, t, n) => (oi(e, typeof t != "symbol" ? t + "" : t, n), n);
import { inject as ii, defineComponent as Vn, ref as X, reactive as ri, computed as le, watch as V, onBeforeUnmount as Kn, onMounted as Jn, openBlock as Gt, createElementBlock as de, normalizeClass as hn, normalizeStyle as Qn, renderSlot as Ce, unref as Me, createCommentVNode as si, createElementVNode as Zn, nextTick as mt, onBeforeMount as ai, provide as ci, withDirectives as li, createVNode as ui, mergeProps as gn, vShow as fi, Fragment as di, renderList as pi, createBlock as hi, withCtx as gi } from "vue";
const to = (e) => !!(e && e.Window) && e instanceof e.Window;
let eo, St;
function no(e) {
  eo = e;
  const t = e.document.createTextNode("");
  t.ownerDocument !== e.document && typeof e.wrap == "function" && e.wrap(t) === t && (e = e.wrap(e)), St = e;
}
typeof window < "u" && window && no(window);
function gt(e) {
  return to(e) ? e : (e.ownerDocument || e).defaultView || St.window;
}
const mi = (e) => e === St || to(e), vi = (e) => be(e) && e.nodeType === 11, be = (e) => !!e && typeof e == "object", oo = (e) => typeof e == "function", yi = (e) => typeof e == "number", bi = (e) => typeof e == "boolean", xi = (e) => typeof e == "string", wi = (e) => {
  if (!e || typeof e != "object")
    return !1;
  const t = gt(e) || St;
  return /object|function/.test(typeof Element) ? e instanceof Element || e instanceof t.Element : e.nodeType === 1 && typeof e.nodeName == "string";
}, Si = (e) => be(e) && !!e.constructor && /function Object\b/.test(e.constructor.toString()), Ei = (e) => be(e) && typeof e.length < "u" && oo(e.splice), g = {
  window: mi,
  docFrag: vi,
  object: be,
  func: oo,
  number: yi,
  bool: bi,
  string: xi,
  element: wi,
  plainObject: Si,
  array: Ei
};
function Ii(e) {
  const {
    actions: t,
    Interactable: n,
    defaults: o
  } = e;
  n.prototype.draggable = ue.draggable, t.map.drag = ue, t.methodDict.drag = "draggable", o.actions.drag = ue.defaults;
}
function ze({
  interaction: e
}) {
  if (e.prepared.name !== "drag")
    return;
  const t = e.prepared.axis;
  t === "x" ? (e.coords.cur.page.y = e.coords.start.page.y, e.coords.cur.client.y = e.coords.start.client.y, e.coords.velocity.client.y = 0, e.coords.velocity.page.y = 0) : t === "y" && (e.coords.cur.page.x = e.coords.start.page.x, e.coords.cur.client.x = e.coords.start.client.x, e.coords.velocity.client.x = 0, e.coords.velocity.page.x = 0);
}
function mn({
  iEvent: e,
  interaction: t
}) {
  if (t.prepared.name !== "drag")
    return;
  const n = t.prepared.axis;
  if (n === "x" || n === "y") {
    const o = n === "x" ? "y" : "x";
    e.page[o] = t.coords.start.page[o], e.client[o] = t.coords.start.client[o], e.delta[o] = 0;
  }
}
const zi = function(t) {
  return g.object(t) ? (this.options.drag.enabled = t.enabled !== !1, this.setPerAction("drag", t), this.setOnEvents("drag", t), /^(xy|x|y|start)$/.test(t.lockAxis) && (this.options.drag.lockAxis = t.lockAxis), /^(xy|x|y)$/.test(t.startAxis) && (this.options.drag.startAxis = t.startAxis), this) : g.bool(t) ? (this.options.drag.enabled = t, this) : this.options.drag;
}, ue = {
  id: "actions/drag",
  install: Ii,
  listeners: {
    "interactions:before-action-move": ze,
    "interactions:action-resume": ze,
    // dragmove
    "interactions:action-move": mn,
    "auto-start:check": (e) => {
      const {
        interaction: t,
        interactable: n,
        buttons: o
      } = e, i = n.options.drag;
      if (!(!(i && i.enabled) || // check mouseButton setting if the pointer is down
      t.pointerIsDown && /mouse|pointer/.test(t.pointerType) && !(o & n.options.drag.mouseButtons)))
        return e.action = {
          name: "drag",
          axis: i.lockAxis === "start" ? i.startAxis : i.lockAxis
        }, !1;
    }
  },
  draggable: zi,
  beforeMove: ze,
  move: mn,
  defaults: {
    startAxis: "xy",
    lockAxis: "xy"
  },
  getCursor() {
    return "move";
  }
}, io = ue, ot = {
  init: Ti,
  document: null,
  DocumentFragment: null,
  SVGElement: null,
  SVGSVGElement: null,
  SVGElementInstance: null,
  Element: null,
  HTMLElement: null,
  Event: null,
  Touch: null,
  PointerEvent: null
};
function $t() {
}
const Q = ot;
function Ti(e) {
  const t = e;
  ot.document = t.document, ot.DocumentFragment = t.DocumentFragment || $t, ot.SVGElement = t.SVGElement || $t, ot.SVGSVGElement = t.SVGSVGElement || $t, ot.SVGElementInstance = t.SVGElementInstance || $t, ot.Element = t.Element || $t, ot.HTMLElement = t.HTMLElement || ot.Element, ot.Event = t.Event, ot.Touch = t.Touch || $t, ot.PointerEvent = t.PointerEvent || t.MSPointerEvent;
}
const it = {
  init: Pi,
  supportsTouch: null,
  supportsPointerEvent: null,
  isIOS7: null,
  isIOS: null,
  isIe9: null,
  isOperaMobile: null,
  prefixedMatchesSelector: null,
  pEventTypes: null,
  wheelEvent: null
};
function Pi(e) {
  const t = Q.Element, n = e.navigator || {};
  it.supportsTouch = "ontouchstart" in e || g.func(e.DocumentTouch) && Q.document instanceof e.DocumentTouch, it.supportsPointerEvent = n.pointerEnabled !== !1 && !!Q.PointerEvent, it.isIOS = /iP(hone|od|ad)/.test(n.platform), it.isIOS7 = /iP(hone|od|ad)/.test(n.platform) && /OS 7[^\d]/.test(n.appVersion), it.isIe9 = /MSIE 9/.test(n.userAgent), it.isOperaMobile = n.appName === "Opera" && it.supportsTouch && /Presto/.test(n.userAgent), it.prefixedMatchesSelector = "matches" in t.prototype ? "matches" : "webkitMatchesSelector" in t.prototype ? "webkitMatchesSelector" : "mozMatchesSelector" in t.prototype ? "mozMatchesSelector" : "oMatchesSelector" in t.prototype ? "oMatchesSelector" : "msMatchesSelector", it.pEventTypes = it.supportsPointerEvent ? Q.PointerEvent === e.MSPointerEvent ? {
    up: "MSPointerUp",
    down: "MSPointerDown",
    over: "mouseover",
    out: "mouseout",
    move: "MSPointerMove",
    cancel: "MSPointerCancel"
  } : {
    up: "pointerup",
    down: "pointerdown",
    over: "pointerover",
    out: "pointerout",
    move: "pointermove",
    cancel: "pointercancel"
  } : null, it.wheelEvent = Q.document && "onmousewheel" in Q.document ? "mousewheel" : "wheel";
}
const rt = it;
function Ot(e, t) {
  if (e.contains)
    return e.contains(t);
  for (; t; ) {
    if (t === e)
      return !0;
    t = t.parentNode;
  }
  return !1;
}
function ro(e, t) {
  for (; g.element(e); ) {
    if (wt(e, t))
      return e;
    e = at(e);
  }
  return null;
}
function at(e) {
  let t = e.parentNode;
  if (g.docFrag(t)) {
    for (; (t = t.host) && g.docFrag(t); )
      ;
    return t;
  }
  return t;
}
function wt(e, t) {
  return St !== eo && (t = t.replace(/\/deep\//g, " ")), e[rt.prefixedMatchesSelector](t);
}
const Re = (e) => e.parentNode || e.host;
function _i(e) {
  let t = [], n;
  for (let o = 0; o < e.length; o++) {
    const i = e[o], r = e[n];
    if (!i || o === n)
      continue;
    if (!r) {
      n = o;
      continue;
    }
    const s = Re(i), a = Re(r);
    if (s === i.ownerDocument)
      continue;
    if (a === i.ownerDocument) {
      n = o;
      continue;
    }
    if (s === a) {
      Di(i, r) && (n = o);
      continue;
    }
    t = t.length ? t : vn(r);
    let l;
    if (r instanceof Q.HTMLElement && i instanceof Q.SVGElement && !(i instanceof Q.SVGSVGElement)) {
      if (i === a)
        continue;
      l = i.ownerSVGElement;
    } else
      l = i;
    const c = vn(l, r.ownerDocument);
    let u = 0;
    for (; c[u] && c[u] === t[u]; )
      u++;
    const f = [c[u - 1], c[u], t[u]];
    if (f[0]) {
      let h = f[0].lastChild;
      for (; h; ) {
        if (h === f[1]) {
          n = o, t = c;
          break;
        } else if (h === f[2])
          break;
        h = h.previousSibling;
      }
    }
  }
  return n;
}
function vn(e, t) {
  const n = [];
  let o = e, i;
  for (; (i = Re(o)) && o !== t && i !== o.ownerDocument; )
    n.unshift(o), o = i;
  return n;
}
function Di(e, t) {
  const n = parseInt(gt(e).getComputedStyle(e).zIndex, 10) || 0, o = parseInt(gt(t).getComputedStyle(t).zIndex, 10) || 0;
  return n >= o;
}
function ke(e, t, n) {
  for (; g.element(e); ) {
    if (wt(e, t))
      return !0;
    if (e = at(e), e === n)
      return wt(e, t);
  }
  return !1;
}
function yn(e) {
  return e.correspondingUseElement || e;
}
function Oi(e) {
  return e = e || St, {
    x: e.scrollX || e.document.documentElement.scrollLeft,
    y: e.scrollY || e.document.documentElement.scrollTop
  };
}
function Ze(e) {
  const t = e instanceof Q.SVGElement ? e.getBoundingClientRect() : e.getClientRects()[0];
  return t && {
    left: t.left,
    right: t.right,
    top: t.top,
    bottom: t.bottom,
    width: t.width || t.right - t.left,
    height: t.height || t.bottom - t.top
  };
}
function tn(e) {
  const t = Ze(e);
  if (!rt.isIOS7 && t) {
    const n = Oi(gt(e));
    t.left += n.x, t.right += n.x, t.top += n.y, t.bottom += n.y;
  }
  return t;
}
function so(e) {
  const t = [];
  for (; e; )
    t.push(e), e = at(e);
  return t;
}
function bn(e) {
  return g.string(e) ? (Q.document.querySelector(e), !0) : !1;
}
function P(e, t) {
  for (const o in t)
    e[o] = t[o];
  return e;
}
function ao(e, t, n) {
  return e === "parent" ? at(n) : e === "self" ? t.getRect(n) : ro(n, e);
}
function Jt(e, t, n, o) {
  let i = e;
  return g.string(i) ? i = ao(i, t, n) : g.func(i) && (i = i(...o)), g.element(i) && (i = tn(i)), i;
}
function xe(e) {
  return e && {
    x: "x" in e ? e.x : e.left,
    y: "y" in e ? e.y : e.top
  };
}
function Ci(e) {
  return e && !("left" in e && "top" in e) && (e = P({}, e), e.left = e.x || 0, e.top = e.y || 0, e.right = e.right || e.left + e.width, e.bottom = e.bottom || e.top + e.height), e;
}
function Ae(e) {
  return e && !("x" in e && "y" in e) && (e = P({}, e), e.x = e.left || 0, e.y = e.top || 0, e.width = e.width || (e.right || 0) - e.x, e.height = e.height || (e.bottom || 0) - e.y), e;
}
function we(e, t, n) {
  e.left && (t.left += n.x), e.right && (t.right += n.x), e.top && (t.top += n.y), e.bottom && (t.bottom += n.y), t.width = t.right - t.left, t.height = t.bottom - t.top;
}
function te(e, t, n) {
  const o = n && e.options[n], r = o && o.origin || e.options.origin, s = Jt(r, e, t, [e && t]);
  return xe(s) || {
    x: 0,
    y: 0
  };
}
function Ct(e, t, n) {
  if (n = n || {}, g.string(e) && e.search(" ") !== -1 && (e = xn(e)), g.array(e))
    return e.reduce((o, i) => P(o, Ct(i, t, n)), n);
  if (g.object(e) && (t = e, e = ""), g.func(t))
    n[e] = n[e] || [], n[e].push(t);
  else if (g.array(t))
    for (const o of t)
      Ct(e, o, n);
  else if (g.object(t))
    for (const o in t) {
      const i = xn(o).map((r) => `${e}${r}`);
      Ct(i, t[o], n);
    }
  return n;
}
function xn(e) {
  return e.trim().split(/ +/);
}
const ee = (e, t) => Math.sqrt(e * e + t * t), Mi = ["webkit", "moz"];
function pe(e, t) {
  e.__set || (e.__set = {});
  for (const n in t)
    Mi.some((o) => n.indexOf(o) === 0) || typeof e[n] != "function" && n !== "__set" && Object.defineProperty(e, n, {
      get() {
        return n in e.__set ? e.__set[n] : e.__set[n] = t[n];
      },
      set(o) {
        e.__set[n] = o;
      },
      configurable: !0
    });
  return e;
}
function fe(e, t) {
  e.page = e.page || {}, e.page.x = t.page.x, e.page.y = t.page.y, e.client = e.client || {}, e.client.x = t.client.x, e.client.y = t.client.y, e.timeStamp = t.timeStamp;
}
function Ri(e, t, n) {
  e.page.x = n.page.x - t.page.x, e.page.y = n.page.y - t.page.y, e.client.x = n.client.x - t.client.x, e.client.y = n.client.y - t.client.y, e.timeStamp = n.timeStamp - t.timeStamp;
}
function ki(e, t) {
  const n = Math.max(t.timeStamp / 1e3, 1e-3);
  e.page.x = t.page.x / n, e.page.y = t.page.y / n, e.client.x = t.client.x / n, e.client.y = t.client.y / n, e.timeStamp = n;
}
function co(e) {
  e.page.x = 0, e.page.y = 0, e.client.x = 0, e.client.y = 0;
}
function lo(e) {
  return e instanceof Q.Event || e instanceof Q.Touch;
}
function he(e, t, n) {
  return n = n || {}, e = e || "page", n.x = t[e + "X"], n.y = t[e + "Y"], n;
}
function uo(e, t) {
  return t = t || {
    x: 0,
    y: 0
  }, rt.isOperaMobile && lo(e) ? (he("screen", e, t), t.x += window.scrollX, t.y += window.scrollY) : he("page", e, t), t;
}
function Ai(e, t) {
  return t = t || {}, rt.isOperaMobile && lo(e) ? he("screen", e, t) : he("client", e, t), t;
}
function Qt(e) {
  return g.number(e.pointerId) ? e.pointerId : e.identifier;
}
function Ni(e, t, n) {
  const o = t.length > 1 ? fo(t) : t[0];
  uo(o, e.page), Ai(o, e.client), e.timeStamp = n;
}
function en(e) {
  const t = [];
  return g.array(e) ? (t[0] = e[0], t[1] = e[1]) : e.type === "touchend" ? e.touches.length === 1 ? (t[0] = e.touches[0], t[1] = e.changedTouches[0]) : e.touches.length === 0 && (t[0] = e.changedTouches[0], t[1] = e.changedTouches[1]) : (t[0] = e.touches[0], t[1] = e.touches[1]), t;
}
function fo(e) {
  const t = {
    pageX: 0,
    pageY: 0,
    clientX: 0,
    clientY: 0,
    screenX: 0,
    screenY: 0
  };
  for (const n of e)
    for (const o in t)
      t[o] += n[o];
  for (const n in t)
    t[n] /= e.length;
  return t;
}
function Ne(e) {
  if (!e.length)
    return null;
  const t = en(e), n = Math.min(t[0].pageX, t[1].pageX), o = Math.min(t[0].pageY, t[1].pageY), i = Math.max(t[0].pageX, t[1].pageX), r = Math.max(t[0].pageY, t[1].pageY);
  return {
    x: n,
    y: o,
    left: n,
    top: o,
    right: i,
    bottom: r,
    width: i - n,
    height: r - o
  };
}
function $e(e, t) {
  const n = t + "X", o = t + "Y", i = en(e), r = i[0][n] - i[1][n], s = i[0][o] - i[1][o];
  return ee(r, s);
}
function Le(e, t) {
  const n = t + "X", o = t + "Y", i = en(e), r = i[1][n] - i[0][n], s = i[1][o] - i[0][o];
  return 180 * Math.atan2(s, r) / Math.PI;
}
function po(e) {
  return g.string(e.pointerType) ? e.pointerType : g.number(e.pointerType) ? [void 0, void 0, "touch", "pen", "mouse"][e.pointerType] : (
    // if the PointerEvent API isn't available, then the "pointer" must
    // be either a MouseEvent, TouchEvent, or Touch object
    /touch/.test(e.type || "") || e instanceof Q.Touch ? "touch" : "mouse"
  );
}
function ho(e) {
  const t = g.func(e.composedPath) ? e.composedPath() : e.path;
  return [yn(t ? t[0] : e.target), yn(e.currentTarget)];
}
function jt() {
  return {
    page: {
      x: 0,
      y: 0
    },
    client: {
      x: 0,
      y: 0
    },
    timeStamp: 0
  };
}
function $i(e) {
  return {
    coords: e,
    get page() {
      return this.coords.page;
    },
    get client() {
      return this.coords.client;
    },
    get timeStamp() {
      return this.coords.timeStamp;
    },
    get pageX() {
      return this.coords.page.x;
    },
    get pageY() {
      return this.coords.page.y;
    },
    get clientX() {
      return this.coords.client.x;
    },
    get clientY() {
      return this.coords.client.y;
    },
    get pointerId() {
      return this.coords.pointerId;
    },
    get target() {
      return this.coords.target;
    },
    get type() {
      return this.coords.type;
    },
    get pointerType() {
      return this.coords.pointerType;
    },
    get buttons() {
      return this.coords.buttons;
    },
    preventDefault() {
    }
  };
}
class Se {
  constructor(t) {
    d(this, "immediatePropagationStopped", !1);
    d(this, "propagationStopped", !1);
    this._interaction = t;
  }
  preventDefault() {
  }
  /**
   * Don't call any other listeners (even on the current target)
   */
  stopPropagation() {
    this.propagationStopped = !0;
  }
  /**
   * Don't call listeners on the remaining targets
   */
  stopImmediatePropagation() {
    this.immediatePropagationStopped = this.propagationStopped = !0;
  }
}
Object.defineProperty(Se.prototype, "interaction", {
  get() {
    return this._interaction._proxy;
  },
  set() {
  }
});
const Li = (e, t) => e.indexOf(t) !== -1, Hi = (e, t) => e.splice(e.indexOf(t), 1), go = (e, t) => {
  for (const n of t)
    e.push(n);
  return e;
}, nn = (e) => go([], e), ne = (e, t) => {
  for (let n = 0; n < e.length; n++)
    if (t(e[n], n, e))
      return n;
  return -1;
}, ge = (e, t) => e[ne(e, t)];
class pt extends Se {
  /**
   * Class of events fired on dropzones during drags with acceptable targets.
   */
  constructor(n, o, i) {
    super(o._interaction);
    d(this, "dropzone");
    d(this, "dragEvent");
    d(this, "relatedTarget");
    d(this, "draggable");
    d(this, "propagationStopped", !1);
    d(this, "immediatePropagationStopped", !1);
    const {
      element: r,
      dropzone: s
    } = i === "dragleave" ? n.prev : n.cur;
    this.type = i, this.target = r, this.currentTarget = r, this.dropzone = s, this.dragEvent = o, this.relatedTarget = o.target, this.draggable = o.interactable, this.timeStamp = o.timeStamp;
  }
  /**
   * If this is a `dropactivate` event, the dropzone element will be
   * deactivated.
   *
   * If this is a `dragmove` or `dragenter`, a `dragleave` will be fired on the
   * dropzone element and more.
   */
  reject() {
    const {
      dropState: n
    } = this._interaction;
    if (!(this.type !== "dropactivate" && (!this.dropzone || n.cur.dropzone !== this.dropzone || n.cur.element !== this.target)))
      if (n.prev.dropzone = this.dropzone, n.prev.element = this.target, n.rejected = !0, n.events.enter = null, this.stopImmediatePropagation(), this.type === "dropactivate") {
        const o = n.activeDrops, i = ne(o, ({
          dropzone: s,
          element: a
        }) => s === this.dropzone && a === this.target);
        n.activeDrops.splice(i, 1);
        const r = new pt(n, this.dragEvent, "dropdeactivate");
        r.dropzone = this.dropzone, r.target = this.target, this.dropzone.fire(r);
      } else
        this.dropzone.fire(new pt(n, this.dragEvent, "dragleave"));
  }
  preventDefault() {
  }
  stopPropagation() {
    this.propagationStopped = !0;
  }
  stopImmediatePropagation() {
    this.immediatePropagationStopped = this.propagationStopped = !0;
  }
}
function Fi(e) {
  const {
    actions: t,
    /** @lends module:interact */
    interactStatic: n,
    /** @lends Interactable */
    Interactable: o,
    defaults: i
  } = e;
  e.usePlugin(io), o.prototype.dropzone = function(r) {
    return Bi(this, r);
  }, o.prototype.dropCheck = function(r, s, a, l, c, u) {
    return Wi(this, r, s, a, l, c, u);
  }, n.dynamicDrop = function(r) {
    return g.bool(r) ? (e.dynamicDrop = r, n) : e.dynamicDrop;
  }, P(t.phaselessTypes, {
    dragenter: !0,
    dragleave: !0,
    dropactivate: !0,
    dropdeactivate: !0,
    dropmove: !0,
    drop: !0
  }), t.methodDict.drop = "dropzone", e.dynamicDrop = !1, i.actions.drop = yo.defaults;
}
function ji({
  interactables: e
}, t) {
  const n = [];
  for (const o of e.list) {
    if (!o.options.drop.enabled)
      continue;
    const i = o.options.drop.accept;
    if (g.element(i) && i !== t || g.string(i) && !wt(t, i) || g.func(i) && !i({
      dropzone: o,
      draggableElement: t
    }))
      continue;
    const r = g.string(o.target) ? o._context.querySelectorAll(o.target) : g.array(o.target) ? o.target : [o.target];
    for (const s of r)
      s !== t && n.push({
        dropzone: o,
        element: s,
        rect: o.getRect(s)
      });
  }
  return n;
}
function mo(e, t) {
  for (const {
    dropzone: n,
    element: o
  } of e.slice())
    t.dropzone = n, t.target = o, n.fire(t), t.propagationStopped = t.immediatePropagationStopped = !1;
}
function He(e, t) {
  const n = ji(e, t);
  for (const o of n)
    o.rect = o.dropzone.getRect(o.element);
  return n;
}
function vo({
  dropState: e,
  interactable: t,
  element: n
}, o, i) {
  const r = [];
  for (const {
    dropzone: a,
    element: l,
    rect: c
  } of e.activeDrops)
    a.dropCheck(o, i, t, n, l, c) && r.push(l);
  const s = _i(r);
  return e.activeDrops[s] || null;
}
function Fe(e, t, n) {
  const o = e.dropState, i = {
    enter: null,
    leave: null,
    activate: null,
    deactivate: null,
    move: null,
    drop: null
  };
  return n.type === "dragstart" && (i.activate = new pt(o, n, "dropactivate"), i.activate.target = null, i.activate.dropzone = null), n.type === "dragend" && (i.deactivate = new pt(o, n, "dropdeactivate"), i.deactivate.target = null, i.deactivate.dropzone = null), o.rejected || (o.cur.element !== o.prev.element && (o.prev.dropzone && (i.leave = new pt(o, n, "dragleave"), n.dragLeave = i.leave.target = o.prev.element, n.prevDropzone = i.leave.dropzone = o.prev.dropzone), o.cur.dropzone && (i.enter = new pt(o, n, "dragenter"), n.dragEnter = o.cur.element, n.dropzone = o.cur.dropzone)), n.type === "dragend" && o.cur.dropzone && (i.drop = new pt(o, n, "drop"), n.dropzone = o.cur.dropzone, n.relatedTarget = o.cur.element), n.type === "dragmove" && o.cur.dropzone && (i.move = new pt(o, n, "dropmove"), n.dropzone = o.cur.dropzone)), i;
}
function Te(e, t) {
  const n = e.dropState, {
    activeDrops: o,
    cur: i,
    prev: r
  } = n;
  t.leave && r.dropzone.fire(t.leave), t.enter && i.dropzone.fire(t.enter), t.move && i.dropzone.fire(t.move), t.drop && i.dropzone.fire(t.drop), t.deactivate && mo(o, t.deactivate), n.prev.dropzone = i.dropzone, n.prev.element = i.element;
}
function wn({
  interaction: e,
  iEvent: t,
  event: n
}, o) {
  if (t.type !== "dragmove" && t.type !== "dragend")
    return;
  const i = e.dropState;
  o.dynamicDrop && (i.activeDrops = He(o, e.element));
  const r = t, s = vo(e, r, n);
  i.rejected = i.rejected && !!s && s.dropzone === i.cur.dropzone && s.element === i.cur.element, i.cur.dropzone = s && s.dropzone, i.cur.element = s && s.element, i.events = Fe(e, n, r);
}
function Bi(e, t) {
  if (g.object(t)) {
    if (e.options.drop.enabled = t.enabled !== !1, t.listeners) {
      const n = Ct(t.listeners), o = Object.keys(n).reduce((r, s) => {
        const a = /^(enter|leave)/.test(s) ? `drag${s}` : /^(activate|deactivate|move)/.test(s) ? `drop${s}` : s;
        return r[a] = n[s], r;
      }, {}), i = e.options.drop.listeners;
      i && e.off(i), e.on(o), e.options.drop.listeners = o;
    }
    return g.func(t.ondrop) && e.on("drop", t.ondrop), g.func(t.ondropactivate) && e.on("dropactivate", t.ondropactivate), g.func(t.ondropdeactivate) && e.on("dropdeactivate", t.ondropdeactivate), g.func(t.ondragenter) && e.on("dragenter", t.ondragenter), g.func(t.ondragleave) && e.on("dragleave", t.ondragleave), g.func(t.ondropmove) && e.on("dropmove", t.ondropmove), /^(pointer|center)$/.test(t.overlap) ? e.options.drop.overlap = t.overlap : g.number(t.overlap) && (e.options.drop.overlap = Math.max(Math.min(1, t.overlap), 0)), "accept" in t && (e.options.drop.accept = t.accept), "checker" in t && (e.options.drop.checker = t.checker), e;
  }
  return g.bool(t) ? (e.options.drop.enabled = t, e) : e.options.drop;
}
function Wi(e, t, n, o, i, r, s) {
  let a = !1;
  if (!(s = s || e.getRect(r)))
    return e.options.drop.checker ? e.options.drop.checker(t, n, a, e, r, o, i) : !1;
  const l = e.options.drop.overlap;
  if (l === "pointer") {
    const u = te(o, i, "drag"), f = uo(t);
    f.x += u.x, f.y += u.y;
    const h = f.x > s.left && f.x < s.right, m = f.y > s.top && f.y < s.bottom;
    a = h && m;
  }
  const c = o.getRect(i);
  if (c && l === "center") {
    const u = c.left + c.width / 2, f = c.top + c.height / 2;
    a = u >= s.left && u <= s.right && f >= s.top && f <= s.bottom;
  }
  return c && g.number(l) && (a = Math.max(0, Math.min(s.right, c.right) - Math.max(s.left, c.left)) * Math.max(0, Math.min(s.bottom, c.bottom) - Math.max(s.top, c.top)) / (c.width * c.height) >= l), e.options.drop.checker && (a = e.options.drop.checker(t, n, a, e, r, o, i)), a;
}
const yo = {
  id: "actions/drop",
  install: Fi,
  listeners: {
    "interactions:before-action-start": ({
      interaction: e
    }) => {
      e.prepared.name === "drag" && (e.dropState = {
        cur: {
          dropzone: null,
          element: null
        },
        prev: {
          dropzone: null,
          element: null
        },
        rejected: null,
        events: null,
        activeDrops: []
      });
    },
    "interactions:after-action-start": ({
      interaction: e,
      event: t,
      iEvent: n
    }, o) => {
      if (e.prepared.name !== "drag")
        return;
      const i = e.dropState;
      i.activeDrops = [], i.events = {}, i.activeDrops = He(o, e.element), i.events = Fe(e, t, n), i.events.activate && (mo(i.activeDrops, i.events.activate), o.fire("actions/drop:start", {
        interaction: e,
        dragEvent: n
      }));
    },
    "interactions:action-move": wn,
    "interactions:after-action-move": ({
      interaction: e,
      iEvent: t
    }, n) => {
      if (e.prepared.name !== "drag")
        return;
      const o = e.dropState;
      Te(e, o.events), n.fire("actions/drop:move", {
        interaction: e,
        dragEvent: t
      }), o.events = {};
    },
    "interactions:action-end": (e, t) => {
      if (e.interaction.prepared.name !== "drag")
        return;
      const {
        interaction: n,
        iEvent: o
      } = e;
      wn(e, t), Te(n, n.dropState.events), t.fire("actions/drop:end", {
        interaction: n,
        dragEvent: o
      });
    },
    "interactions:stop": ({
      interaction: e
    }) => {
      if (e.prepared.name !== "drag")
        return;
      const {
        dropState: t
      } = e;
      t && (t.activeDrops = null, t.events = null, t.cur.dropzone = null, t.cur.element = null, t.prev.dropzone = null, t.prev.element = null, t.rejected = !1);
    }
  },
  getActiveDrops: He,
  getDrop: vo,
  getDropEvents: Fe,
  fireDropEvents: Te,
  defaults: {
    enabled: !1,
    accept: null,
    overlap: "pointer"
  }
}, Yi = yo;
function Xi(e) {
  const {
    actions: t,
    Interactable: n,
    defaults: o
  } = e;
  n.prototype.gesturable = function(i) {
    return g.object(i) ? (this.options.gesture.enabled = i.enabled !== !1, this.setPerAction("gesture", i), this.setOnEvents("gesture", i), this) : g.bool(i) ? (this.options.gesture.enabled = i, this) : this.options.gesture;
  }, t.map.gesture = je, t.methodDict.gesture = "gesturable", o.actions.gesture = je.defaults;
}
function Pe({
  interaction: e,
  iEvent: t,
  phase: n
}) {
  if (e.prepared.name !== "gesture")
    return;
  const o = e.pointers.map((a) => a.pointer), i = n === "start", r = n === "end", s = e.interactable.options.deltaSource;
  if (t.touches = [o[0], o[1]], i)
    t.distance = $e(o, s), t.box = Ne(o), t.scale = 1, t.ds = 0, t.angle = Le(o, s), t.da = 0, e.gesture.startDistance = t.distance, e.gesture.startAngle = t.angle;
  else if (r) {
    const a = e.prevEvent;
    t.distance = a.distance, t.box = a.box, t.scale = a.scale, t.ds = 0, t.angle = a.angle, t.da = 0;
  } else
    t.distance = $e(o, s), t.box = Ne(o), t.scale = t.distance / e.gesture.startDistance, t.angle = Le(o, s), t.ds = t.scale - e.gesture.scale, t.da = t.angle - e.gesture.angle;
  e.gesture.distance = t.distance, e.gesture.angle = t.angle, g.number(t.scale) && t.scale !== 1 / 0 && !isNaN(t.scale) && (e.gesture.scale = t.scale);
}
const je = {
  id: "actions/gesture",
  before: ["actions/drag", "actions/resize"],
  install: Xi,
  listeners: {
    "interactions:action-start": Pe,
    "interactions:action-move": Pe,
    "interactions:action-end": Pe,
    "interactions:new": ({
      interaction: e
    }) => {
      e.gesture = {
        angle: 0,
        distance: 0,
        scale: 1,
        startAngle: 0,
        startDistance: 0
      };
    },
    "auto-start:check": (e) => {
      if (e.interaction.pointers.length < 2)
        return;
      const t = e.interactable.options.gesture;
      if (t && t.enabled)
        return e.action = {
          name: "gesture"
        }, !1;
    }
  },
  defaults: {},
  getCursor() {
    return "";
  }
}, Gi = je;
function Ui(e) {
  const {
    actions: t,
    browser: n,
    /** @lends Interactable */
    Interactable: o,
    // tslint:disable-line no-shadowed-variable
    defaults: i
  } = e;
  ht.cursors = Ji(n), ht.defaultMargin = n.supportsTouch || n.supportsPointerEvent ? 20 : 10, o.prototype.resizable = function(r) {
    return Vi(this, r, e);
  }, t.map.resize = ht, t.methodDict.resize = "resizable", i.actions.resize = ht.defaults;
}
function qi(e) {
  const {
    interaction: t,
    interactable: n,
    element: o,
    rect: i,
    buttons: r
  } = e;
  if (!i)
    return;
  const s = P({}, t.coords.cur.page), a = n.options.resize;
  if (!(!(a && a.enabled) || // check mouseButton setting if the pointer is down
  t.pointerIsDown && /mouse|pointer/.test(t.pointerType) && !(r & a.mouseButtons))) {
    if (g.object(a.edges)) {
      const l = {
        left: !1,
        right: !1,
        top: !1,
        bottom: !1
      };
      for (const c in l)
        l[c] = Ki(c, a.edges[c], s, t._latestPointer.eventTarget, o, i, a.margin || ht.defaultMargin);
      l.left = l.left && !l.right, l.top = l.top && !l.bottom, (l.left || l.right || l.top || l.bottom) && (e.action = {
        name: "resize",
        edges: l
      });
    } else {
      const l = a.axis !== "y" && s.x > i.right - ht.defaultMargin, c = a.axis !== "x" && s.y > i.bottom - ht.defaultMargin;
      (l || c) && (e.action = {
        name: "resize",
        axes: (l ? "x" : "") + (c ? "y" : "")
      });
    }
    return e.action ? !1 : void 0;
  }
}
function Vi(e, t, n) {
  return g.object(t) ? (e.options.resize.enabled = t.enabled !== !1, e.setPerAction("resize", t), e.setOnEvents("resize", t), g.string(t.axis) && /^x$|^y$|^xy$/.test(t.axis) ? e.options.resize.axis = t.axis : t.axis === null && (e.options.resize.axis = n.defaults.actions.resize.axis), g.bool(t.preserveAspectRatio) ? e.options.resize.preserveAspectRatio = t.preserveAspectRatio : g.bool(t.square) && (e.options.resize.square = t.square), e) : g.bool(t) ? (e.options.resize.enabled = t, e) : e.options.resize;
}
function Ki(e, t, n, o, i, r, s) {
  if (!t)
    return !1;
  if (t === !0) {
    const a = g.number(r.width) ? r.width : r.right - r.left, l = g.number(r.height) ? r.height : r.bottom - r.top;
    if (s = Math.min(s, Math.abs((e === "left" || e === "right" ? a : l) / 2)), a < 0 && (e === "left" ? e = "right" : e === "right" && (e = "left")), l < 0 && (e === "top" ? e = "bottom" : e === "bottom" && (e = "top")), e === "left") {
      const c = a >= 0 ? r.left : r.right;
      return n.x < c + s;
    }
    if (e === "top") {
      const c = l >= 0 ? r.top : r.bottom;
      return n.y < c + s;
    }
    if (e === "right")
      return n.x > (a >= 0 ? r.right : r.left) - s;
    if (e === "bottom")
      return n.y > (l >= 0 ? r.bottom : r.top) - s;
  }
  return g.element(o) ? g.element(t) ? (
    // the value is an element to use as a resize handle
    t === o
  ) : (
    // otherwise check if element matches value as selector
    ke(o, t, i)
  ) : !1;
}
function Ji(e) {
  return e.isIe9 ? {
    x: "e-resize",
    y: "s-resize",
    xy: "se-resize",
    top: "n-resize",
    left: "w-resize",
    bottom: "s-resize",
    right: "e-resize",
    topleft: "se-resize",
    bottomright: "se-resize",
    topright: "ne-resize",
    bottomleft: "ne-resize"
  } : {
    x: "ew-resize",
    y: "ns-resize",
    xy: "nwse-resize",
    top: "ns-resize",
    left: "ew-resize",
    bottom: "ns-resize",
    right: "ew-resize",
    topleft: "nwse-resize",
    bottomright: "nwse-resize",
    topright: "nesw-resize",
    bottomleft: "nesw-resize"
  };
}
function Qi({
  iEvent: e,
  interaction: t
}) {
  if (t.prepared.name !== "resize" || !t.prepared.edges)
    return;
  const n = e, o = t.rect;
  t._rects = {
    start: P({}, o),
    corrected: P({}, o),
    previous: P({}, o),
    delta: {
      left: 0,
      right: 0,
      width: 0,
      top: 0,
      bottom: 0,
      height: 0
    }
  }, n.edges = t.prepared.edges, n.rect = t._rects.corrected, n.deltaRect = t._rects.delta;
}
function Zi({
  iEvent: e,
  interaction: t
}) {
  if (t.prepared.name !== "resize" || !t.prepared.edges)
    return;
  const n = e, i = t.interactable.options.resize.invert, r = i === "reposition" || i === "negate", s = t.rect, {
    start: a,
    corrected: l,
    delta: c,
    previous: u
  } = t._rects;
  if (P(u, l), r) {
    if (P(l, s), i === "reposition") {
      if (l.top > l.bottom) {
        const f = l.top;
        l.top = l.bottom, l.bottom = f;
      }
      if (l.left > l.right) {
        const f = l.left;
        l.left = l.right, l.right = f;
      }
    }
  } else
    l.top = Math.min(s.top, a.bottom), l.bottom = Math.max(s.bottom, a.top), l.left = Math.min(s.left, a.right), l.right = Math.max(s.right, a.left);
  l.width = l.right - l.left, l.height = l.bottom - l.top;
  for (const f in l)
    c[f] = l[f] - u[f];
  n.edges = t.prepared.edges, n.rect = l, n.deltaRect = c;
}
function tr({
  iEvent: e,
  interaction: t
}) {
  if (t.prepared.name !== "resize" || !t.prepared.edges)
    return;
  const n = e;
  n.edges = t.prepared.edges, n.rect = t._rects.corrected, n.deltaRect = t._rects.delta;
}
function Sn({
  iEvent: e,
  interaction: t
}) {
  if (t.prepared.name !== "resize" || !t.resizeAxes)
    return;
  const n = t.interactable.options, o = e;
  n.resize.square ? (t.resizeAxes === "y" ? o.delta.x = o.delta.y : o.delta.y = o.delta.x, o.axes = "xy") : (o.axes = t.resizeAxes, t.resizeAxes === "x" ? o.delta.y = 0 : t.resizeAxes === "y" && (o.delta.x = 0));
}
const ht = {
  id: "actions/resize",
  before: ["actions/drag"],
  install: Ui,
  listeners: {
    "interactions:new": ({
      interaction: e
    }) => {
      e.resizeAxes = "xy";
    },
    "interactions:action-start": (e) => {
      Qi(e), Sn(e);
    },
    "interactions:action-move": (e) => {
      Zi(e), Sn(e);
    },
    "interactions:action-end": tr,
    "auto-start:check": qi
  },
  defaults: {
    square: !1,
    preserveAspectRatio: !1,
    axis: "xy",
    // use default margin
    margin: NaN,
    // object with props left, right, top, bottom which are
    // true/false values to resize when the pointer is over that edge,
    // CSS selectors to match the handles for each direction
    // or the Elements for each handle
    edges: null,
    // a value of 'none' will limit the resize rect to a minimum of 0x0
    // 'negate' will alow the rect to have negative width/height
    // 'reposition' will keep the width/height positive by swapping
    // the top and bottom edges and/or swapping the left and right edges
    invert: "none"
  },
  cursors: null,
  getCursor({
    edges: e,
    axis: t,
    name: n
  }) {
    const o = ht.cursors;
    let i = null;
    if (t)
      i = o[n + t];
    else if (e) {
      let r = "";
      for (const s of ["top", "bottom", "left", "right"])
        e[s] && (r += s);
      i = o[r];
    }
    return i;
  },
  defaultMargin: null
}, er = ht, nr = {
  id: "actions",
  install(e) {
    e.usePlugin(Gi), e.usePlugin(er), e.usePlugin(io), e.usePlugin(Yi);
  }
};
let En = 0, dt, Pt;
function or(e) {
  if (dt = e.requestAnimationFrame, Pt = e.cancelAnimationFrame, !dt) {
    const t = ["ms", "moz", "webkit", "o"];
    for (const n of t)
      dt = e[`${n}RequestAnimationFrame`], Pt = e[`${n}CancelAnimationFrame`] || e[`${n}CancelRequestAnimationFrame`];
  }
  dt = dt && dt.bind(e), Pt = Pt && Pt.bind(e), dt || (dt = (t) => {
    const n = Date.now(), o = Math.max(0, 16 - (n - En)), i = e.setTimeout(() => {
      t(n + o);
    }, o);
    return En = n + o, i;
  }, Pt = (t) => clearTimeout(t));
}
const bt = {
  request: (e) => dt(e),
  cancel: (e) => Pt(e),
  init: or
};
function ir(e) {
  const {
    defaults: t,
    actions: n
  } = e;
  e.autoScroll = O, O.now = () => e.now(), n.phaselessTypes.autoscroll = !0, t.perAction.autoScroll = O.defaults;
}
const O = {
  defaults: {
    enabled: !1,
    margin: 60,
    // the item that is scrolled (Window or HTMLElement)
    container: null,
    // the scroll speed in pixels per second
    speed: 300
  },
  now: Date.now,
  interaction: null,
  i: 0,
  // the handle returned by window.setInterval
  // Direction each pulse is to scroll in
  x: 0,
  y: 0,
  isScrolling: !1,
  prevTime: 0,
  margin: 0,
  speed: 0,
  start(e) {
    O.isScrolling = !0, bt.cancel(O.i), e.autoScroll = O, O.interaction = e, O.prevTime = O.now(), O.i = bt.request(O.scroll);
  },
  stop() {
    O.isScrolling = !1, O.interaction && (O.interaction.autoScroll = null), bt.cancel(O.i);
  },
  // scroll the window by the values in scroll.x/y
  scroll() {
    const {
      interaction: e
    } = O, {
      interactable: t,
      element: n
    } = e, o = e.prepared.name, i = t.options[o].autoScroll, r = In(i.container, t, n), s = O.now(), a = (s - O.prevTime) / 1e3, l = i.speed * a;
    if (l >= 1) {
      const c = {
        x: O.x * l,
        y: O.y * l
      };
      if (c.x || c.y) {
        const u = zn(r);
        g.window(r) ? r.scrollBy(c.x, c.y) : r && (r.scrollLeft += c.x, r.scrollTop += c.y);
        const f = zn(r), h = {
          x: f.x - u.x,
          y: f.y - u.y
        };
        (h.x || h.y) && t.fire({
          type: "autoscroll",
          target: n,
          interactable: t,
          delta: h,
          interaction: e,
          container: r
        });
      }
      O.prevTime = s;
    }
    O.isScrolling && (bt.cancel(O.i), O.i = bt.request(O.scroll));
  },
  check(e, t) {
    var n;
    return (n = e.options[t].autoScroll) == null ? void 0 : n.enabled;
  },
  onInteractionMove({
    interaction: e,
    pointer: t
  }) {
    if (!(e.interacting() && O.check(e.interactable, e.prepared.name)))
      return;
    if (e.simulation) {
      O.x = O.y = 0;
      return;
    }
    let n, o, i, r;
    const {
      interactable: s,
      element: a
    } = e, l = e.prepared.name, c = s.options[l].autoScroll, u = In(c.container, s, a);
    if (g.window(u))
      r = t.clientX < O.margin, n = t.clientY < O.margin, o = t.clientX > u.innerWidth - O.margin, i = t.clientY > u.innerHeight - O.margin;
    else {
      const f = Ze(u);
      r = t.clientX < f.left + O.margin, n = t.clientY < f.top + O.margin, o = t.clientX > f.right - O.margin, i = t.clientY > f.bottom - O.margin;
    }
    O.x = o ? 1 : r ? -1 : 0, O.y = i ? 1 : n ? -1 : 0, O.isScrolling || (O.margin = c.margin, O.speed = c.speed, O.start(e));
  }
};
function In(e, t, n) {
  return (g.string(e) ? ao(e, t, n) : e) || gt(n);
}
function zn(e) {
  return g.window(e) && (e = window.document.body), {
    x: e.scrollLeft,
    y: e.scrollTop
  };
}
const rr = {
  id: "auto-scroll",
  install: ir,
  listeners: {
    "interactions:new": ({
      interaction: e
    }) => {
      e.autoScroll = null;
    },
    "interactions:destroy": ({
      interaction: e
    }) => {
      e.autoScroll = null, O.stop(), O.interaction && (O.interaction = null);
    },
    "interactions:stop": O.stop,
    "interactions:action-move": (e) => O.onInteractionMove(e)
  }
}, sr = rr;
function Zt(e, t) {
  let n = !1;
  return function() {
    return n || (St.console.warn(t), n = !0), e.apply(this, arguments);
  };
}
function on(e, t) {
  return e.name = t.name, e.axis = t.axis, e.edges = t.edges, e;
}
function ar(e) {
  const {
    /** @lends Interactable */
    Interactable: t
    // tslint:disable-line no-shadowed-variable
  } = e;
  t.prototype.getAction = function(o, i, r, s) {
    const a = cr(this, i, r, s, e);
    return this.options.actionChecker ? this.options.actionChecker(o, i, a, this, s, r) : a;
  }, t.prototype.ignoreFrom = Zt(function(n) {
    return this._backCompatOption("ignoreFrom", n);
  }, "Interactable.ignoreFrom() has been deprecated. Use Interactble.draggable({ignoreFrom: newValue})."), t.prototype.allowFrom = Zt(function(n) {
    return this._backCompatOption("allowFrom", n);
  }, "Interactable.allowFrom() has been deprecated. Use Interactble.draggable({allowFrom: newValue})."), t.prototype.actionChecker = ur, t.prototype.styleCursor = lr;
}
function cr(e, t, n, o, i) {
  const r = e.getRect(o), s = t.buttons || {
    0: 1,
    1: 4,
    3: 8,
    4: 16
  }[t.button], a = {
    action: null,
    interactable: e,
    interaction: n,
    element: o,
    rect: r,
    buttons: s
  };
  return i.fire("auto-start:check", a), a.action;
}
function lr(e) {
  return g.bool(e) ? (this.options.styleCursor = e, this) : e === null ? (delete this.options.styleCursor, this) : this.options.styleCursor;
}
function ur(e) {
  return g.func(e) ? (this.options.actionChecker = e, this) : e === null ? (delete this.options.actionChecker, this) : this.options.actionChecker;
}
const fr = {
  id: "auto-start/interactableMethods",
  install: ar
};
function dr(e) {
  const {
    interactStatic: t,
    defaults: n
  } = e;
  e.usePlugin(fr), n.base.actionChecker = null, n.base.styleCursor = !0, P(n.perAction, {
    manualStart: !1,
    max: 1 / 0,
    maxPerElement: 1,
    allowFrom: null,
    ignoreFrom: null,
    // only allow left button by default
    // see https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/buttons#Return_value
    mouseButtons: 1
  }), t.maxInteractions = (o) => So(o, e), e.autoStart = {
    // Allow this many interactions to happen simultaneously
    maxInteractions: 1 / 0,
    withinInteractionLimit: Ee,
    cursorElement: null
  };
}
function pr({
  interaction: e,
  pointer: t,
  event: n,
  eventTarget: o
}, i) {
  if (e.interacting())
    return;
  const r = xo(e, t, n, o, i);
  wo(e, r, i);
}
function hr({
  interaction: e,
  pointer: t,
  event: n,
  eventTarget: o
}, i) {
  if (e.pointerType !== "mouse" || e.pointerIsDown || e.interacting())
    return;
  const r = xo(e, t, n, o, i);
  wo(e, r, i);
}
function gr(e, t) {
  const {
    interaction: n
  } = e;
  if (!n.pointerIsDown || n.interacting() || !n.pointerWasMoved || !n.prepared.name)
    return;
  t.fire("autoStart:before-start", e);
  const {
    interactable: o
  } = n, i = n.prepared.name;
  i && o && (o.options[i].manualStart || !Ee(o, n.element, n.prepared, t) ? n.stop() : (n.start(n.prepared, o, n.element), Eo(n, t)));
}
function mr({
  interaction: e
}, t) {
  const {
    interactable: n
  } = e;
  n && n.options.styleCursor && Be(e.element, "", t);
}
function bo(e, t, n, o, i) {
  return t.testIgnoreAllow(t.options[e.name], n, o) && t.options[e.name].enabled && Ee(t, n, e, i) ? e : null;
}
function vr(e, t, n, o, i, r, s) {
  for (let a = 0, l = o.length; a < l; a++) {
    const c = o[a], u = i[a], f = c.getAction(t, n, e, u);
    if (!f)
      continue;
    const h = bo(f, c, u, r, s);
    if (h)
      return {
        action: h,
        interactable: c,
        element: u
      };
  }
  return {
    action: null,
    interactable: null,
    element: null
  };
}
function xo(e, t, n, o, i) {
  let r = [], s = [], a = o;
  function l(c) {
    r.push(c), s.push(a);
  }
  for (; g.element(a); ) {
    r = [], s = [], i.interactables.forEachMatch(a, l);
    const c = vr(e, t, n, r, s, o, i);
    if (c.action && !c.interactable.options[c.action.name].manualStart)
      return c;
    a = at(a);
  }
  return {
    action: null,
    interactable: null,
    element: null
  };
}
function wo(e, {
  action: t,
  interactable: n,
  element: o
}, i) {
  t = t || {
    name: null
  }, e.interactable = n, e.element = o, on(e.prepared, t), e.rect = n && t.name ? n.getRect(o) : null, Eo(e, i), i.fire("autoStart:prepared", {
    interaction: e
  });
}
function Ee(e, t, n, o) {
  const i = e.options, r = i[n.name].max, s = i[n.name].maxPerElement, a = o.autoStart.maxInteractions;
  let l = 0, c = 0, u = 0;
  if (!(r && s && a))
    return !1;
  for (const f of o.interactions.list) {
    const h = f.prepared.name;
    if (f.interacting()) {
      if (l++, l >= a)
        return !1;
      if (f.interactable === e && (c += h === n.name ? 1 : 0, c >= r || f.element === t && (u++, h === n.name && u >= s)))
        return !1;
    }
  }
  return a > 0;
}
function So(e, t) {
  return g.number(e) ? (t.autoStart.maxInteractions = e, this) : t.autoStart.maxInteractions;
}
function Be(e, t, n) {
  const {
    cursorElement: o
  } = n.autoStart;
  o && o !== e && (o.style.cursor = ""), e.ownerDocument.documentElement.style.cursor = t, e.style.cursor = t, n.autoStart.cursorElement = t ? e : null;
}
function Eo(e, t) {
  const {
    interactable: n,
    element: o,
    prepared: i
  } = e;
  if (!(e.pointerType === "mouse" && n && n.options.styleCursor)) {
    t.autoStart.cursorElement && Be(t.autoStart.cursorElement, "", t);
    return;
  }
  let r = "";
  if (i.name) {
    const s = n.options[i.name].cursorChecker;
    g.func(s) ? r = s(i, n, o, e._interacting) : r = t.actions.map[i.name].getCursor(i);
  }
  Be(e.element, r || "", t);
}
const yr = {
  id: "auto-start/base",
  before: ["actions"],
  install: dr,
  listeners: {
    "interactions:down": pr,
    "interactions:move": (e, t) => {
      hr(e, t), gr(e, t);
    },
    "interactions:stop": mr
  },
  maxInteractions: So,
  withinInteractionLimit: Ee,
  validateAction: bo
}, rn = yr;
function br({
  interaction: e,
  eventTarget: t,
  dx: n,
  dy: o
}, i) {
  if (e.prepared.name !== "drag")
    return;
  const r = Math.abs(n), s = Math.abs(o), a = e.interactable.options.drag, l = a.startAxis, c = r > s ? "x" : r < s ? "y" : "xy";
  if (e.prepared.axis = a.lockAxis === "start" ? c[0] : a.lockAxis, c !== "xy" && l !== "xy" && l !== c) {
    e.prepared.name = null;
    let u = t;
    const f = function(h) {
      if (h === e.interactable)
        return;
      const m = e.interactable.options.drag;
      if (!m.manualStart && h.testIgnoreAllow(m, u, t)) {
        const v = h.getAction(e.downPointer, e.downEvent, e, u);
        if (v && v.name === "drag" && xr(c, h) && rn.validateAction(v, h, u, t, i))
          return h;
      }
    };
    for (; g.element(u); ) {
      const h = i.interactables.forEachMatch(u, f);
      if (h) {
        e.prepared.name = "drag", e.interactable = h, e.element = u;
        break;
      }
      u = at(u);
    }
  }
}
function xr(e, t) {
  if (!t)
    return !1;
  const n = t.options.drag.startAxis;
  return e === "xy" || n === "xy" || n === e;
}
const wr = {
  id: "auto-start/dragAxis",
  listeners: {
    "autoStart:before-start": br
  }
};
function Sr(e) {
  const {
    defaults: t
  } = e;
  e.usePlugin(rn), t.perAction.hold = 0, t.perAction.delay = 0;
}
function _e(e) {
  const t = e.prepared && e.prepared.name;
  if (!t)
    return null;
  const n = e.interactable.options;
  return n[t].hold || n[t].delay;
}
const Er = {
  id: "auto-start/hold",
  install: Sr,
  listeners: {
    "interactions:new": ({
      interaction: e
    }) => {
      e.autoStartHoldTimer = null;
    },
    "autoStart:prepared": ({
      interaction: e
    }) => {
      const t = _e(e);
      t > 0 && (e.autoStartHoldTimer = setTimeout(() => {
        e.start(e.prepared, e.interactable, e.element);
      }, t));
    },
    "interactions:move": ({
      interaction: e,
      duplicate: t
    }) => {
      e.autoStartHoldTimer && e.pointerWasMoved && !t && (clearTimeout(e.autoStartHoldTimer), e.autoStartHoldTimer = null);
    },
    // prevent regular down->move autoStart
    "autoStart:before-start": ({
      interaction: e
    }) => {
      _e(e) > 0 && (e.prepared.name = null);
    }
  },
  getHoldDuration: _e
}, Ir = Er, zr = {
  id: "auto-start",
  install(e) {
    e.usePlugin(rn), e.usePlugin(Ir), e.usePlugin(wr);
  }
};
function Tr(e) {
  return /^(always|never|auto)$/.test(e) ? (this.options.preventDefault = e, this) : g.bool(e) ? (this.options.preventDefault = e ? "always" : "never", this) : this.options.preventDefault;
}
function Pr(e, t, n) {
  const o = e.options.preventDefault;
  if (o !== "never") {
    if (o === "always") {
      n.preventDefault();
      return;
    }
    if (t.events.supportsPassive && /^touch(start|move)$/.test(n.type)) {
      const i = gt(n.target).document, r = t.getDocOptions(i);
      if (!(r && r.events) || r.events.passive !== !1)
        return;
    }
    /^(mouse|pointer|touch)*(down|start)/i.test(n.type) || g.element(n.target) && wt(n.target, "input,select,textarea,[contenteditable=true],[contenteditable=true] *") || n.preventDefault();
  }
}
function _r({
  interaction: e,
  event: t
}) {
  e.interactable && e.interactable.checkAndPreventDefault(t);
}
function Dr(e) {
  const {
    Interactable: t
  } = e;
  t.prototype.preventDefault = Tr, t.prototype.checkAndPreventDefault = function(n) {
    return Pr(this, e, n);
  }, e.interactions.docEvents.push({
    type: "dragstart",
    listener(n) {
      for (const o of e.interactions.list)
        if (o.element && (o.element === n.target || Ot(o.element, n.target))) {
          o.interactable.checkAndPreventDefault(n);
          return;
        }
    }
  });
}
const Io = {
  id: "core/interactablePreventDefault",
  install: Dr,
  listeners: ["down", "move", "up", "cancel"].reduce((e, t) => (e[`interactions:${t}`] = _r, e), {})
}, Or = {};
var Lt;
(function(e) {
  e.touchAction = "touchAction", e.boxSizing = "boxSizing", e.noListeners = "noListeners";
})(Lt || (Lt = {}));
const Tn = "[interact.js] ", We = {
  touchAction: "https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action",
  boxSizing: "https://developer.mozilla.org/en-US/docs/Web/CSS/box-sizing"
}, Cr = !1;
function Mr(e, {
  logger: t
} = {}) {
  const {
    Interactable: n,
    defaults: o
  } = e;
  e.logger = t || console, o.base.devTools = {
    ignore: {}
  }, n.prototype.devTools = function(i) {
    return i ? (P(this.options.devTools, i), this) : this.options.devTools;
  }, e.usePlugin(Or);
}
const Pn = [{
  name: Lt.touchAction,
  perform({
    element: e
  }) {
    return !!e && !Rr(e, "touchAction", /pan-|pinch|none/);
  },
  getInfo({
    element: e
  }) {
    return [e, We.touchAction];
  },
  text: `Consider adding CSS "touch-action: none" to this element
`
}, {
  name: Lt.boxSizing,
  perform(e) {
    const {
      element: t
    } = e;
    return e.prepared.name === "resize" && t instanceof Q.HTMLElement && !zo(t, "boxSizing", /border-box/);
  },
  text: 'Consider adding CSS "box-sizing: border-box" to this resizable element',
  getInfo({
    element: e
  }) {
    return [e, We.boxSizing];
  }
}, {
  name: Lt.noListeners,
  perform(e) {
    var t;
    const n = e.prepared.name;
    return !(((t = e.interactable) == null ? void 0 : t.events.types[`${n}move`]) || []).length;
  },
  getInfo(e) {
    return [e.prepared.name, e.interactable];
  },
  text: "There are no listeners set for this action"
}];
function zo(e, t, n) {
  const o = e.style[t] || St.getComputedStyle(e)[t];
  return n.test((o || "").toString());
}
function Rr(e, t, n) {
  let o = e;
  for (; g.element(o); ) {
    if (zo(o, t, n))
      return !0;
    o = at(o);
  }
  return !1;
}
const _n = "dev-tools", kr = Cr ? {
  id: _n,
  install: () => {
  }
} : {
  id: _n,
  install: Mr,
  listeners: {
    "interactions:action-start": ({
      interaction: e
    }, t) => {
      for (const n of Pn) {
        const o = e.interactable && e.interactable.options;
        !(o && o.devTools && o.devTools.ignore[n.name]) && n.perform(e) && t.logger.warn(Tn + n.text, ...n.getInfo(e));
      }
    }
  },
  checks: Pn,
  CheckName: Lt,
  links: We,
  prefix: Tn
}, Ar = kr;
function Ht(e) {
  const t = {};
  for (const n in e) {
    const o = e[n];
    g.plainObject(o) ? t[n] = Ht(o) : g.array(o) ? t[n] = nn(o) : t[n] = o;
  }
  return t;
}
class sn {
  constructor(t) {
    d(this, "states", []);
    d(this, "startOffset", {
      left: 0,
      right: 0,
      top: 0,
      bottom: 0
    });
    d(this, "startDelta");
    d(this, "result");
    d(this, "endResult");
    d(this, "startEdges");
    d(this, "edges");
    d(this, "interaction");
    this.interaction = t, this.result = se(), this.edges = {
      left: !1,
      right: !1,
      top: !1,
      bottom: !1
    };
  }
  start({
    phase: t
  }, n) {
    const {
      interaction: o
    } = this, i = Nr(o);
    this.prepareStates(i), this.startEdges = P({}, o.edges), this.edges = P({}, this.startEdges), this.startOffset = $r(o.rect, n), this.startDelta = {
      x: 0,
      y: 0
    };
    const r = this.fillArg({
      phase: t,
      pageCoords: n,
      preEnd: !1
    });
    return this.result = se(), this.startAll(r), this.result = this.setAll(r);
  }
  fillArg(t) {
    const {
      interaction: n
    } = this;
    return t.interaction = n, t.interactable = n.interactable, t.element = n.element, t.rect || (t.rect = n.rect), t.edges || (t.edges = this.startEdges), t.startOffset = this.startOffset, t;
  }
  startAll(t) {
    for (const n of this.states)
      n.methods.start && (t.state = n, n.methods.start(t));
  }
  setAll(t) {
    const {
      phase: n,
      preEnd: o,
      skipModifiers: i,
      rect: r,
      edges: s
    } = t;
    t.coords = P({}, t.pageCoords), t.rect = P({}, r), t.edges = P({}, s);
    const a = i ? this.states.slice(i) : this.states, l = se(t.coords, t.rect);
    for (const h of a) {
      var c;
      const {
        options: m
      } = h, v = P({}, t.coords);
      let D = null;
      (c = h.methods) != null && c.set && this.shouldDo(m, o, n) && (t.state = h, D = h.methods.set(t), we(t.edges, t.rect, {
        x: t.coords.x - v.x,
        y: t.coords.y - v.y
      })), l.eventProps.push(D);
    }
    P(this.edges, t.edges), l.delta.x = t.coords.x - t.pageCoords.x, l.delta.y = t.coords.y - t.pageCoords.y, l.rectDelta.left = t.rect.left - r.left, l.rectDelta.right = t.rect.right - r.right, l.rectDelta.top = t.rect.top - r.top, l.rectDelta.bottom = t.rect.bottom - r.bottom;
    const u = this.result.coords, f = this.result.rect;
    if (u && f) {
      const h = l.rect.left !== f.left || l.rect.right !== f.right || l.rect.top !== f.top || l.rect.bottom !== f.bottom;
      l.changed = h || u.x !== l.coords.x || u.y !== l.coords.y;
    }
    return l;
  }
  applyToInteraction(t) {
    const {
      interaction: n
    } = this, {
      phase: o
    } = t, i = n.coords.cur, r = n.coords.start, {
      result: s,
      startDelta: a
    } = this, l = s.delta;
    o === "start" && P(this.startDelta, s.delta);
    for (const [f, h] of [[r, a], [i, l]])
      f.page.x += h.x, f.page.y += h.y, f.client.x += h.x, f.client.y += h.y;
    const {
      rectDelta: c
    } = this.result, u = t.rect || n.rect;
    u.left += c.left, u.right += c.right, u.top += c.top, u.bottom += c.bottom, u.width = u.right - u.left, u.height = u.bottom - u.top;
  }
  setAndApply(t) {
    const {
      interaction: n
    } = this, {
      phase: o,
      preEnd: i,
      skipModifiers: r
    } = t, s = this.setAll(this.fillArg({
      preEnd: i,
      phase: o,
      pageCoords: t.modifiedCoords || n.coords.cur.page
    }));
    if (this.result = s, !s.changed && (!r || r < this.states.length) && n.interacting())
      return !1;
    if (t.modifiedCoords) {
      const {
        page: a
      } = n.coords.cur, l = {
        x: t.modifiedCoords.x - a.x,
        y: t.modifiedCoords.y - a.y
      };
      s.coords.x += l.x, s.coords.y += l.y, s.delta.x += l.x, s.delta.y += l.y;
    }
    this.applyToInteraction(t);
  }
  beforeEnd(t) {
    const {
      interaction: n,
      event: o
    } = t, i = this.states;
    if (!i || !i.length)
      return;
    let r = !1;
    for (const s of i) {
      t.state = s;
      const {
        options: a,
        methods: l
      } = s, c = l.beforeEnd && l.beforeEnd(t);
      if (c)
        return this.endResult = c, !1;
      r = r || !r && this.shouldDo(a, !0, t.phase, !0);
    }
    r && n.move({
      event: o,
      preEnd: !0
    });
  }
  stop(t) {
    const {
      interaction: n
    } = t;
    if (!this.states || !this.states.length)
      return;
    const o = P({
      states: this.states,
      interactable: n.interactable,
      element: n.element,
      rect: null
    }, t);
    this.fillArg(o);
    for (const i of this.states)
      o.state = i, i.methods.stop && i.methods.stop(o);
    this.states = null, this.endResult = null;
  }
  prepareStates(t) {
    this.states = [];
    for (let n = 0; n < t.length; n++) {
      const {
        options: o,
        methods: i,
        name: r
      } = t[n];
      this.states.push({
        options: o,
        methods: i,
        index: n,
        name: r
      });
    }
    return this.states;
  }
  restoreInteractionCoords({
    interaction: {
      coords: t,
      rect: n,
      modification: o
    }
  }) {
    if (!o.result)
      return;
    const {
      startDelta: i
    } = o, {
      delta: r,
      rectDelta: s
    } = o.result, a = [[t.start, i], [t.cur, r]];
    for (const [l, c] of a)
      l.page.x -= c.x, l.page.y -= c.y, l.client.x -= c.x, l.client.y -= c.y;
    n.left -= s.left, n.right -= s.right, n.top -= s.top, n.bottom -= s.bottom;
  }
  shouldDo(t, n, o, i) {
    return (
      // ignore disabled modifiers
      !(!t || t.enabled === !1 || // check if we require endOnly option to fire move before end
      i && !t.endOnly || // don't apply endOnly modifiers when not ending
      t.endOnly && !n || // check if modifier should run be applied on start
      o === "start" && !t.setStart)
    );
  }
  copyFrom(t) {
    this.startOffset = t.startOffset, this.startDelta = t.startDelta, this.startEdges = t.startEdges, this.edges = t.edges, this.states = t.states.map((n) => Ht(n)), this.result = se(P({}, t.result.coords), P({}, t.result.rect));
  }
  destroy() {
    for (const t in this)
      this[t] = null;
  }
}
function se(e, t) {
  return {
    rect: t,
    coords: e,
    delta: {
      x: 0,
      y: 0
    },
    rectDelta: {
      left: 0,
      right: 0,
      top: 0,
      bottom: 0
    },
    eventProps: [],
    changed: !0
  };
}
function Nr(e) {
  const t = e.interactable.options[e.prepared.name], n = t.modifiers;
  return n && n.length ? n : ["snap", "snapSize", "snapEdges", "restrict", "restrictEdges", "restrictSize"].map((o) => {
    const i = t[o];
    return i && i.enabled && {
      options: i,
      methods: i._methods
    };
  }).filter((o) => !!o);
}
function $r(e, t) {
  return e ? {
    left: t.x - e.left,
    top: t.y - e.top,
    right: e.right - t.x,
    bottom: e.bottom - t.y
  } : {
    left: 0,
    top: 0,
    right: 0,
    bottom: 0
  };
}
function Et(e, t) {
  const {
    defaults: n
  } = e, o = {
    start: e.start,
    set: e.set,
    beforeEnd: e.beforeEnd,
    stop: e.stop
  }, i = (r) => {
    const s = r || {};
    s.enabled = s.enabled !== !1;
    for (const l in n)
      l in s || (s[l] = n[l]);
    const a = {
      options: s,
      methods: o,
      name: t,
      enable: () => (s.enabled = !0, a),
      disable: () => (s.enabled = !1, a)
    };
    return a;
  };
  return t && typeof t == "string" && (i._defaults = n, i._methods = o), i;
}
function Ut({
  iEvent: e,
  interaction: t
}) {
  const n = t.modification.result;
  n && (e.modifiers = n.eventProps);
}
const Lr = {
  id: "modifiers/base",
  before: ["actions"],
  install: (e) => {
    e.defaults.perAction.modifiers = [];
  },
  listeners: {
    "interactions:new": ({
      interaction: e
    }) => {
      e.modification = new sn(e);
    },
    "interactions:before-action-start": (e) => {
      const {
        interaction: t
      } = e, n = e.interaction.modification;
      n.start(e, t.coords.start.page), t.edges = n.edges, n.applyToInteraction(e);
    },
    "interactions:before-action-move": (e) => {
      const {
        interaction: t
      } = e, {
        modification: n
      } = t, o = n.setAndApply(e);
      return t.edges = n.edges, o;
    },
    "interactions:before-action-end": (e) => {
      const {
        interaction: t
      } = e, {
        modification: n
      } = t, o = n.beforeEnd(e);
      return t.edges = n.startEdges, o;
    },
    "interactions:action-start": Ut,
    "interactions:action-move": Ut,
    "interactions:action-end": Ut,
    "interactions:after-action-start": (e) => e.interaction.modification.restoreInteractionCoords(e),
    "interactions:after-action-move": (e) => e.interaction.modification.restoreInteractionCoords(e),
    "interactions:stop": (e) => e.interaction.modification.stop(e)
  }
}, To = Lr, Po = {
  base: {
    preventDefault: "auto",
    deltaSource: "page"
  },
  perAction: {
    enabled: !1,
    origin: {
      x: 0,
      y: 0
    }
  },
  actions: {}
};
class an extends Se {
  /** */
  constructor(n, o, i, r, s, a, l) {
    super(n);
    d(this, "relatedTarget", null);
    d(this, "screenX");
    d(this, "screenY");
    d(this, "button");
    d(this, "buttons");
    d(this, "ctrlKey");
    d(this, "shiftKey");
    d(this, "altKey");
    d(this, "metaKey");
    d(this, "page");
    d(this, "client");
    d(this, "delta");
    d(this, "rect");
    d(this, "x0");
    d(this, "y0");
    d(this, "t0");
    d(this, "dt");
    d(this, "duration");
    d(this, "clientX0");
    d(this, "clientY0");
    d(this, "velocity");
    d(this, "speed");
    d(this, "swipe");
    // resize
    d(this, "axes");
    d(this, "preEnd");
    s = s || n.element;
    const c = n.interactable, u = (c && c.options || Po).deltaSource, f = te(c, s, i), h = r === "start", m = r === "end", v = h ? this : n.prevEvent, D = h ? n.coords.start : m ? {
      page: v.page,
      client: v.client,
      timeStamp: n.coords.cur.timeStamp
    } : n.coords.cur;
    this.page = P({}, D.page), this.client = P({}, D.client), this.rect = P({}, n.rect), this.timeStamp = D.timeStamp, m || (this.page.x -= f.x, this.page.y -= f.y, this.client.x -= f.x, this.client.y -= f.y), this.ctrlKey = o.ctrlKey, this.altKey = o.altKey, this.shiftKey = o.shiftKey, this.metaKey = o.metaKey, this.button = o.button, this.buttons = o.buttons, this.target = s, this.currentTarget = s, this.preEnd = a, this.type = l || i + (r || ""), this.interactable = c, this.t0 = h ? n.pointers[n.pointers.length - 1].downTime : v.t0, this.x0 = n.coords.start.page.x - f.x, this.y0 = n.coords.start.page.y - f.y, this.clientX0 = n.coords.start.client.x - f.x, this.clientY0 = n.coords.start.client.y - f.y, h || m ? this.delta = {
      x: 0,
      y: 0
    } : this.delta = {
      x: this[u].x - v[u].x,
      y: this[u].y - v[u].y
    }, this.dt = n.coords.delta.timeStamp, this.duration = this.timeStamp - this.t0, this.velocity = P({}, n.coords.velocity[u]), this.speed = ee(this.velocity.x, this.velocity.y), this.swipe = m || r === "inertiastart" ? this.getSwipe() : null;
  }
  getSwipe() {
    const n = this._interaction;
    if (n.prevEvent.speed < 600 || this.timeStamp - n.prevEvent.timeStamp > 150)
      return null;
    let o = 180 * Math.atan2(n.prevEvent.velocityY, n.prevEvent.velocityX) / Math.PI;
    const i = 22.5;
    o < 0 && (o += 360);
    const r = 135 - i <= o && o < 225 + i, s = 225 - i <= o && o < 315 + i, a = !r && (315 - i <= o || o < 45 + i), l = !s && 45 - i <= o && o < 135 + i;
    return {
      up: s,
      down: l,
      left: r,
      right: a,
      angle: o,
      speed: n.prevEvent.speed,
      velocity: {
        x: n.prevEvent.velocityX,
        y: n.prevEvent.velocityY
      }
    };
  }
  preventDefault() {
  }
  /**
   * Don't call listeners on the remaining targets
   */
  stopImmediatePropagation() {
    this.immediatePropagationStopped = this.propagationStopped = !0;
  }
  /**
   * Don't call any other listeners (even on the current target)
   */
  stopPropagation() {
    this.propagationStopped = !0;
  }
}
Object.defineProperties(an.prototype, {
  pageX: {
    get() {
      return this.page.x;
    },
    set(e) {
      this.page.x = e;
    }
  },
  pageY: {
    get() {
      return this.page.y;
    },
    set(e) {
      this.page.y = e;
    }
  },
  clientX: {
    get() {
      return this.client.x;
    },
    set(e) {
      this.client.x = e;
    }
  },
  clientY: {
    get() {
      return this.client.y;
    },
    set(e) {
      this.client.y = e;
    }
  },
  dx: {
    get() {
      return this.delta.x;
    },
    set(e) {
      this.delta.x = e;
    }
  },
  dy: {
    get() {
      return this.delta.y;
    },
    set(e) {
      this.delta.y = e;
    }
  },
  velocityX: {
    get() {
      return this.velocity.x;
    },
    set(e) {
      this.velocity.x = e;
    }
  },
  velocityY: {
    get() {
      return this.velocity.y;
    },
    set(e) {
      this.velocity.y = e;
    }
  }
});
class Hr {
  constructor(t, n, o, i, r) {
    d(this, "id");
    d(this, "pointer");
    d(this, "event");
    d(this, "downTime");
    d(this, "downTarget");
    this.id = t, this.pointer = n, this.event = o, this.downTime = i, this.downTarget = r;
  }
}
let Ye;
(function(e) {
  e.interactable = "", e.element = "", e.prepared = "", e.pointerIsDown = "", e.pointerWasMoved = "", e._proxy = "";
})(Ye || (Ye = {}));
let me;
(function(e) {
  e.start = "", e.move = "", e.end = "", e.stop = "", e.interacting = "";
})(me || (me = {}));
let Fr = 0;
class jr {
  /** */
  constructor({
    pointerType: t,
    scopeFire: n
  }) {
    // current interactable being interacted with
    d(this, "interactable", null);
    // the target element of the interactable
    d(this, "element", null);
    d(this, "rect", null);
    d(this, "_rects");
    d(this, "edges", null);
    d(this, "_scopeFire");
    // action that's ready to be fired on next move event
    d(this, "prepared", {
      name: null,
      axis: null,
      edges: null
    });
    d(this, "pointerType");
    // keep track of added pointers
    d(this, "pointers", []);
    // pointerdown/mousedown/touchstart event
    d(this, "downEvent", null);
    d(this, "downPointer", {});
    d(this, "_latestPointer", {
      pointer: null,
      event: null,
      eventTarget: null
    });
    // previous action event
    d(this, "prevEvent", null);
    d(this, "pointerIsDown", !1);
    d(this, "pointerWasMoved", !1);
    d(this, "_interacting", !1);
    d(this, "_ending", !1);
    d(this, "_stopped", !0);
    d(this, "_proxy");
    d(this, "simulation", null);
    /**
     * @alias Interaction.prototype.move
     */
    d(this, "doMove", Zt(function(t) {
      this.move(t);
    }, "The interaction.doMove() method has been renamed to interaction.move()"));
    d(this, "coords", {
      // Starting InteractEvent pointer coordinates
      start: jt(),
      // Previous native pointer move event coordinates
      prev: jt(),
      // current native pointer move event coordinates
      cur: jt(),
      // Change in coordinates and time of the pointer
      delta: jt(),
      // pointer velocity
      velocity: jt()
    });
    d(this, "_id", Fr++);
    this._scopeFire = n, this.pointerType = t;
    const o = this;
    this._proxy = {};
    for (const i in Ye)
      Object.defineProperty(this._proxy, i, {
        get() {
          return o[i];
        }
      });
    for (const i in me)
      Object.defineProperty(this._proxy, i, {
        value: (...r) => o[i](...r)
      });
    this._scopeFire("interactions:new", {
      interaction: this
    });
  }
  /** @internal */
  get pointerMoveTolerance() {
    return 1;
  }
  pointerDown(t, n, o) {
    const i = this.updatePointer(t, n, o, !0), r = this.pointers[i];
    this._scopeFire("interactions:down", {
      pointer: t,
      event: n,
      eventTarget: o,
      pointerIndex: i,
      pointerInfo: r,
      type: "down",
      interaction: this
    });
  }
  /**
   * ```js
   * interact(target)
   *   .draggable({
   *     // disable the default drag start by down->move
   *     manualStart: true
   *   })
   *   // start dragging after the user holds the pointer down
   *   .on('hold', function (event) {
   *     var interaction = event.interaction
   *
   *     if (!interaction.interacting()) {
   *       interaction.start({ name: 'drag' },
   *                         event.interactable,
   *                         event.currentTarget)
   *     }
   * })
   * ```
   *
   * Start an action with the given Interactable and Element as tartgets. The
   * action must be enabled for the target Interactable and an appropriate
   * number of pointers must be held down - 1 for drag/resize, 2 for gesture.
   *
   * Use it with `interactable.<action>able({ manualStart: false })` to always
   * [start actions manually](https://github.com/taye/interact.js/issues/114)
   *
   * @param {object} action   The action to be performed - drag, resize, etc.
   * @param {Interactable} target  The Interactable to target
   * @param {Element} element The DOM Element to target
   * @return {Boolean} Whether the interaction was successfully started
   */
  start(t, n, o) {
    return this.interacting() || !this.pointerIsDown || this.pointers.length < (t.name === "gesture" ? 2 : 1) || !n.options[t.name].enabled ? !1 : (on(this.prepared, t), this.interactable = n, this.element = o, this.rect = n.getRect(o), this.edges = this.prepared.edges ? P({}, this.prepared.edges) : {
      left: !0,
      right: !0,
      top: !0,
      bottom: !0
    }, this._stopped = !1, this._interacting = this._doPhase({
      interaction: this,
      event: this.downEvent,
      phase: "start"
    }) && !this._stopped, this._interacting);
  }
  pointerMove(t, n, o) {
    !this.simulation && !(this.modification && this.modification.endResult) && this.updatePointer(t, n, o, !1);
    const i = this.coords.cur.page.x === this.coords.prev.page.x && this.coords.cur.page.y === this.coords.prev.page.y && this.coords.cur.client.x === this.coords.prev.client.x && this.coords.cur.client.y === this.coords.prev.client.y;
    let r, s;
    this.pointerIsDown && !this.pointerWasMoved && (r = this.coords.cur.client.x - this.coords.start.client.x, s = this.coords.cur.client.y - this.coords.start.client.y, this.pointerWasMoved = ee(r, s) > this.pointerMoveTolerance);
    const a = this.getPointerIndex(t), l = {
      pointer: t,
      pointerIndex: a,
      pointerInfo: this.pointers[a],
      event: n,
      type: "move",
      eventTarget: o,
      dx: r,
      dy: s,
      duplicate: i,
      interaction: this
    };
    i || ki(this.coords.velocity, this.coords.delta), this._scopeFire("interactions:move", l), !i && !this.simulation && (this.interacting() && (l.type = null, this.move(l)), this.pointerWasMoved && fe(this.coords.prev, this.coords.cur));
  }
  /**
   * ```js
   * interact(target)
   *   .draggable(true)
   *   .on('dragmove', function (event) {
   *     if (someCondition) {
   *       // change the snap settings
   *       event.interactable.draggable({ snap: { targets: [] }})
   *       // fire another move event with re-calculated snap
   *       event.interaction.move()
   *     }
   *   })
   * ```
   *
   * Force a move of the current action at the same coordinates. Useful if
   * snap/restrict has been changed and you want a movement with the new
   * settings.
   */
  move(t) {
    (!t || !t.event) && co(this.coords.delta), t = P({
      pointer: this._latestPointer.pointer,
      event: this._latestPointer.event,
      eventTarget: this._latestPointer.eventTarget,
      interaction: this
    }, t || {}), t.phase = "move", this._doPhase(t);
  }
  // End interact move events and stop auto-scroll unless simulation is running
  pointerUp(t, n, o, i) {
    let r = this.getPointerIndex(t);
    r === -1 && (r = this.updatePointer(t, n, o, !1));
    const s = /cancel$/i.test(n.type) ? "cancel" : "up";
    this._scopeFire(`interactions:${s}`, {
      pointer: t,
      pointerIndex: r,
      pointerInfo: this.pointers[r],
      event: n,
      eventTarget: o,
      type: s,
      curEventTarget: i,
      interaction: this
    }), this.simulation || this.end(n), this.removePointer(t, n);
  }
  documentBlur(t) {
    this.end(t), this._scopeFire("interactions:blur", {
      event: t,
      type: "blur",
      interaction: this
    });
  }
  /**
   * ```js
   * interact(target)
   *   .draggable(true)
   *   .on('move', function (event) {
   *     if (event.pageX > 1000) {
   *       // end the current action
   *       event.interaction.end()
   *       // stop all further listeners from being called
   *       event.stopImmediatePropagation()
   *     }
   *   })
   * ```
   *
   * @param {PointerEvent} [event]
   */
  end(t) {
    this._ending = !0, t = t || this._latestPointer.event;
    let n;
    this.interacting() && (n = this._doPhase({
      event: t,
      interaction: this,
      phase: "end"
    })), this._ending = !1, n === !0 && this.stop();
  }
  currentAction() {
    return this._interacting ? this.prepared.name : null;
  }
  interacting() {
    return this._interacting;
  }
  /** */
  stop() {
    this._scopeFire("interactions:stop", {
      interaction: this
    }), this.interactable = this.element = null, this._interacting = !1, this._stopped = !0, this.prepared.name = this.prevEvent = null;
  }
  getPointerIndex(t) {
    const n = Qt(t);
    return this.pointerType === "mouse" || this.pointerType === "pen" ? this.pointers.length - 1 : ne(this.pointers, (o) => o.id === n);
  }
  getPointerInfo(t) {
    return this.pointers[this.getPointerIndex(t)];
  }
  updatePointer(t, n, o, i) {
    const r = Qt(t);
    let s = this.getPointerIndex(t), a = this.pointers[s];
    return i = i === !1 ? !1 : i || /(down|start)$/i.test(n.type), a ? a.pointer = t : (a = new Hr(r, t, n, null, null), s = this.pointers.length, this.pointers.push(a)), Ni(this.coords.cur, this.pointers.map((l) => l.pointer), this._now()), Ri(this.coords.delta, this.coords.prev, this.coords.cur), i && (this.pointerIsDown = !0, a.downTime = this.coords.cur.timeStamp, a.downTarget = o, pe(this.downPointer, t), this.interacting() || (fe(this.coords.start, this.coords.cur), fe(this.coords.prev, this.coords.cur), this.downEvent = n, this.pointerWasMoved = !1)), this._updateLatestPointer(t, n, o), this._scopeFire("interactions:update-pointer", {
      pointer: t,
      event: n,
      eventTarget: o,
      down: i,
      pointerInfo: a,
      pointerIndex: s,
      interaction: this
    }), s;
  }
  removePointer(t, n) {
    const o = this.getPointerIndex(t);
    if (o === -1)
      return;
    const i = this.pointers[o];
    this._scopeFire("interactions:remove-pointer", {
      pointer: t,
      event: n,
      eventTarget: null,
      pointerIndex: o,
      pointerInfo: i,
      interaction: this
    }), this.pointers.splice(o, 1), this.pointerIsDown = !1;
  }
  _updateLatestPointer(t, n, o) {
    this._latestPointer.pointer = t, this._latestPointer.event = n, this._latestPointer.eventTarget = o;
  }
  destroy() {
    this._latestPointer.pointer = null, this._latestPointer.event = null, this._latestPointer.eventTarget = null;
  }
  _createPreparedEvent(t, n, o, i) {
    return new an(this, t, this.prepared.name, n, this.element, o, i);
  }
  _fireEvent(t) {
    var n;
    (n = this.interactable) == null || n.fire(t), (!this.prevEvent || t.timeStamp >= this.prevEvent.timeStamp) && (this.prevEvent = t);
  }
  _doPhase(t) {
    const {
      event: n,
      phase: o,
      preEnd: i,
      type: r
    } = t, {
      rect: s
    } = this;
    if (s && o === "move" && (we(this.edges, s, this.coords.delta[this.interactable.options.deltaSource]), s.width = s.right - s.left, s.height = s.bottom - s.top), this._scopeFire(`interactions:before-action-${o}`, t) === !1)
      return !1;
    const l = t.iEvent = this._createPreparedEvent(n, o, i, r);
    return this._scopeFire(`interactions:action-${o}`, t), o === "start" && (this.prevEvent = l), this._fireEvent(l), this._scopeFire(`interactions:after-action-${o}`, t), !0;
  }
  _now() {
    return Date.now();
  }
}
me.offsetBy = "";
function Br(e) {
  e.pointerIsDown && (Xe(e.coords.cur, e.offset.total), e.offset.pending.x = 0, e.offset.pending.y = 0);
}
function Dn({
  interaction: e
}) {
  _o(e);
}
function Wr({
  interaction: e
}) {
  if (_o(e))
    return e.move({
      offset: !0
    }), e.end(), !1;
}
function Yr({
  interaction: e
}) {
  e.offset.total.x = 0, e.offset.total.y = 0, e.offset.pending.x = 0, e.offset.pending.y = 0;
}
function _o(e) {
  if (!Gr(e))
    return !1;
  const {
    pending: t
  } = e.offset;
  return Xe(e.coords.cur, t), Xe(e.coords.delta, t), we(e.edges, e.rect, t), t.x = 0, t.y = 0, !0;
}
function Xr({
  x: e,
  y: t
}) {
  this.offset.pending.x += e, this.offset.pending.y += t, this.offset.total.x += e, this.offset.total.y += t;
}
function Xe({
  page: e,
  client: t
}, {
  x: n,
  y: o
}) {
  e.x += n, e.y += o, t.x += n, t.y += o;
}
function Gr(e) {
  return !!(e.offset.pending.x || e.offset.pending.y);
}
const Ur = {
  id: "offset",
  before: ["modifiers", "pointer-events", "actions", "inertia"],
  install(e) {
    e.Interaction.prototype.offsetBy = Xr;
  },
  listeners: {
    "interactions:new": ({
      interaction: e
    }) => {
      e.offset = {
        total: {
          x: 0,
          y: 0
        },
        pending: {
          x: 0,
          y: 0
        }
      };
    },
    "interactions:update-pointer": ({
      interaction: e
    }) => Br(e),
    "interactions:before-action-start": Dn,
    "interactions:before-action-move": Dn,
    "interactions:before-action-end": Wr,
    "interactions:stop": Yr
  }
}, Do = Ur;
function qr(e) {
  const {
    defaults: t
  } = e;
  e.usePlugin(Do), e.usePlugin(To), e.actions.phases.inertiastart = !0, e.actions.phases.resume = !0, t.perAction.inertia = {
    enabled: !1,
    resistance: 10,
    // the lambda in exponential decay
    minSpeed: 100,
    // target speed must be above this for inertia to start
    endSpeed: 10,
    // the speed at which inertia is slow enough to stop
    allowResume: !0,
    // allow resuming an action in inertia phase
    smoothEndDuration: 300
    // animate to snap/restrict endOnly if there's no inertia
  };
}
class Vr {
  constructor(t) {
    d(this, "active", !1);
    d(this, "isModified", !1);
    d(this, "smoothEnd", !1);
    d(this, "allowResume", !1);
    d(this, "modification");
    d(this, "modifierCount", 0);
    d(this, "modifierArg");
    d(this, "startCoords");
    d(this, "t0", 0);
    d(this, "v0", 0);
    d(this, "te", 0);
    d(this, "targetOffset");
    d(this, "modifiedOffset");
    d(this, "currentOffset");
    d(this, "lambda_v0", 0);
    // eslint-disable-line camelcase
    d(this, "one_ve_v0", 0);
    // eslint-disable-line camelcase
    d(this, "timeout");
    d(this, "interaction");
    this.interaction = t;
  }
  start(t) {
    const {
      interaction: n
    } = this, o = ae(n);
    if (!o || !o.enabled)
      return !1;
    const {
      client: i
    } = n.coords.velocity, r = ee(i.x, i.y), s = this.modification || (this.modification = new sn(n));
    if (s.copyFrom(n.modification), this.t0 = n._now(), this.allowResume = o.allowResume, this.v0 = r, this.currentOffset = {
      x: 0,
      y: 0
    }, this.startCoords = n.coords.cur.page, this.modifierArg = s.fillArg({
      pageCoords: this.startCoords,
      preEnd: !0,
      phase: "inertiastart"
    }), this.t0 - n.coords.cur.timeStamp < 50 && r > o.minSpeed && r > o.endSpeed)
      this.startInertia();
    else {
      if (s.result = s.setAll(this.modifierArg), !s.result.changed)
        return !1;
      this.startSmoothEnd();
    }
    return n.modification.result.rect = null, n.offsetBy(this.targetOffset), n._doPhase({
      interaction: n,
      event: t,
      phase: "inertiastart"
    }), n.offsetBy({
      x: -this.targetOffset.x,
      y: -this.targetOffset.y
    }), n.modification.result.rect = null, this.active = !0, n.simulation = this, !0;
  }
  startInertia() {
    const t = this.interaction.coords.velocity.client, n = ae(this.interaction), o = n.resistance, i = -Math.log(n.endSpeed / this.v0) / o;
    this.targetOffset = {
      x: (t.x - i) / o,
      y: (t.y - i) / o
    }, this.te = i, this.lambda_v0 = o / this.v0, this.one_ve_v0 = 1 - n.endSpeed / this.v0;
    const {
      modification: r,
      modifierArg: s
    } = this;
    s.pageCoords = {
      x: this.startCoords.x + this.targetOffset.x,
      y: this.startCoords.y + this.targetOffset.y
    }, r.result = r.setAll(s), r.result.changed && (this.isModified = !0, this.modifiedOffset = {
      x: this.targetOffset.x + r.result.delta.x,
      y: this.targetOffset.y + r.result.delta.y
    }), this.onNextFrame(() => this.inertiaTick());
  }
  startSmoothEnd() {
    this.smoothEnd = !0, this.isModified = !0, this.targetOffset = {
      x: this.modification.result.delta.x,
      y: this.modification.result.delta.y
    }, this.onNextFrame(() => this.smoothEndTick());
  }
  onNextFrame(t) {
    this.timeout = bt.request(() => {
      this.active && t();
    });
  }
  inertiaTick() {
    const {
      interaction: t
    } = this, o = ae(t).resistance, i = (t._now() - this.t0) / 1e3;
    if (i < this.te) {
      const r = 1 - (Math.exp(-o * i) - this.lambda_v0) / this.one_ve_v0;
      let s;
      this.isModified ? s = ts(0, 0, this.targetOffset.x, this.targetOffset.y, this.modifiedOffset.x, this.modifiedOffset.y, r) : s = {
        x: this.targetOffset.x * r,
        y: this.targetOffset.y * r
      };
      const a = {
        x: s.x - this.currentOffset.x,
        y: s.y - this.currentOffset.y
      };
      this.currentOffset.x += a.x, this.currentOffset.y += a.y, t.offsetBy(a), t.move(), this.onNextFrame(() => this.inertiaTick());
    } else
      t.offsetBy({
        x: this.modifiedOffset.x - this.currentOffset.x,
        y: this.modifiedOffset.y - this.currentOffset.y
      }), this.end();
  }
  smoothEndTick() {
    const {
      interaction: t
    } = this, n = t._now() - this.t0, {
      smoothEndDuration: o
    } = ae(t);
    if (n < o) {
      const i = {
        x: Cn(n, 0, this.targetOffset.x, o),
        y: Cn(n, 0, this.targetOffset.y, o)
      }, r = {
        x: i.x - this.currentOffset.x,
        y: i.y - this.currentOffset.y
      };
      this.currentOffset.x += r.x, this.currentOffset.y += r.y, t.offsetBy(r), t.move({
        skipModifiers: this.modifierCount
      }), this.onNextFrame(() => this.smoothEndTick());
    } else
      t.offsetBy({
        x: this.targetOffset.x - this.currentOffset.x,
        y: this.targetOffset.y - this.currentOffset.y
      }), this.end();
  }
  resume({
    pointer: t,
    event: n,
    eventTarget: o
  }) {
    const {
      interaction: i
    } = this;
    i.offsetBy({
      x: -this.currentOffset.x,
      y: -this.currentOffset.y
    }), i.updatePointer(t, n, o, !0), i._doPhase({
      interaction: i,
      event: n,
      phase: "resume"
    }), fe(i.coords.prev, i.coords.cur), this.stop();
  }
  end() {
    this.interaction.move(), this.interaction.end(), this.stop();
  }
  stop() {
    this.active = this.smoothEnd = !1, this.interaction.simulation = null, bt.cancel(this.timeout);
  }
}
function Kr({
  interaction: e,
  event: t
}) {
  return !e._interacting || e.simulation ? null : e.inertia.start(t) ? !1 : null;
}
function Jr(e) {
  const {
    interaction: t,
    eventTarget: n
  } = e, o = t.inertia;
  if (!o.active)
    return;
  let i = n;
  for (; g.element(i); ) {
    if (i === t.element) {
      o.resume(e);
      break;
    }
    i = at(i);
  }
}
function Qr({
  interaction: e
}) {
  const t = e.inertia;
  t.active && t.stop();
}
function ae({
  interactable: e,
  prepared: t
}) {
  return e && e.options && t.name && e.options[t.name].inertia;
}
const Zr = {
  id: "inertia",
  before: ["modifiers", "actions"],
  install: qr,
  listeners: {
    "interactions:new": ({
      interaction: e
    }) => {
      e.inertia = new Vr(e);
    },
    "interactions:before-action-end": Kr,
    "interactions:down": Jr,
    "interactions:stop": Qr,
    "interactions:before-action-resume": (e) => {
      const {
        modification: t
      } = e.interaction;
      t.stop(e), t.start(e, e.interaction.coords.cur.page), t.applyToInteraction(e);
    },
    "interactions:before-action-inertiastart": (e) => e.interaction.modification.setAndApply(e),
    "interactions:action-resume": Ut,
    "interactions:action-inertiastart": Ut,
    "interactions:after-action-inertiastart": (e) => e.interaction.modification.restoreInteractionCoords(e),
    "interactions:after-action-resume": (e) => e.interaction.modification.restoreInteractionCoords(e)
  }
};
function On(e, t, n, o) {
  const i = 1 - e;
  return i * i * t + 2 * i * e * n + e * e * o;
}
function ts(e, t, n, o, i, r, s) {
  return {
    x: On(s, e, n, i),
    y: On(s, t, o, r)
  };
}
function Cn(e, t, n, o) {
  return e /= o, -n * e * (e - 2) + t;
}
const es = Zr;
function Mn(e, t) {
  for (const n of t) {
    if (e.immediatePropagationStopped)
      break;
    n(e);
  }
}
class Oo {
  constructor(t) {
    d(this, "options");
    d(this, "types", {});
    d(this, "propagationStopped", !1);
    d(this, "immediatePropagationStopped", !1);
    d(this, "global");
    this.options = P({}, t || {});
  }
  fire(t) {
    let n;
    const o = this.global;
    (n = this.types[t.type]) && Mn(t, n), !t.propagationStopped && o && (n = o[t.type]) && Mn(t, n);
  }
  on(t, n) {
    const o = Ct(t, n);
    for (t in o)
      this.types[t] = go(this.types[t] || [], o[t]);
  }
  off(t, n) {
    const o = Ct(t, n);
    for (t in o) {
      const i = this.types[t];
      if (!(!i || !i.length))
        for (const r of o[t]) {
          const s = i.indexOf(r);
          s !== -1 && i.splice(s, 1);
        }
    }
  }
  getRect(t) {
    return null;
  }
}
function Ge(e, t) {
  if (t.phaselessTypes[e])
    return !0;
  for (const n in t.map)
    if (e.indexOf(n) === 0 && e.substr(n.length) in t.phases)
      return !0;
  return !1;
}
function ns(e) {
  const t = (n, o) => {
    let i = e.interactables.get(n, o);
    return i || (i = e.interactables.new(n, o), i.events.global = t.globalEvents), i;
  };
  return t.getPointerAverage = fo, t.getTouchBBox = Ne, t.getTouchDistance = $e, t.getTouchAngle = Le, t.getElementRect = tn, t.getElementClientRect = Ze, t.matchesSelector = wt, t.closest = ro, t.globalEvents = {}, t.version = "1.10.20", t.scope = e, t.use = function(n, o) {
    return this.scope.usePlugin(n, o), this;
  }, t.isSet = function(n, o) {
    return !!this.scope.interactables.get(n, o && o.context);
  }, t.on = Zt(function(o, i, r) {
    if (g.string(o) && o.search(" ") !== -1 && (o = o.trim().split(/ +/)), g.array(o)) {
      for (const s of o)
        this.on(s, i, r);
      return this;
    }
    if (g.object(o)) {
      for (const s in o)
        this.on(s, o[s], i);
      return this;
    }
    return Ge(o, this.scope.actions) ? this.globalEvents[o] ? this.globalEvents[o].push(i) : this.globalEvents[o] = [i] : this.scope.events.add(this.scope.document, o, i, {
      options: r
    }), this;
  }, "The interact.on() method is being deprecated"), t.off = Zt(function(o, i, r) {
    if (g.string(o) && o.search(" ") !== -1 && (o = o.trim().split(/ +/)), g.array(o)) {
      for (const s of o)
        this.off(s, i, r);
      return this;
    }
    if (g.object(o)) {
      for (const s in o)
        this.off(s, o[s], i);
      return this;
    }
    if (Ge(o, this.scope.actions)) {
      let s;
      o in this.globalEvents && (s = this.globalEvents[o].indexOf(i)) !== -1 && this.globalEvents[o].splice(s, 1);
    } else
      this.scope.events.remove(this.scope.document, o, i, r);
    return this;
  }, "The interact.off() method is being deprecated"), t.debug = function() {
    return this.scope;
  }, t.supportsTouch = function() {
    return rt.supportsTouch;
  }, t.supportsPointerEvent = function() {
    return rt.supportsPointerEvent;
  }, t.stop = function() {
    for (const n of this.scope.interactions.list)
      n.stop();
    return this;
  }, t.pointerMoveTolerance = function(n) {
    return g.number(n) ? (this.scope.interactions.pointerMoveTolerance = n, this) : this.scope.interactions.pointerMoveTolerance;
  }, t.addDocument = function(n, o) {
    this.scope.addDocument(n, o);
  }, t.removeDocument = function(n) {
    this.scope.removeDocument(n);
  }, t;
}
class os {
  /** */
  constructor(t, n, o, i) {
    d(this, "options");
    d(this, "_actions");
    d(this, "target");
    d(this, "events", new Oo());
    d(this, "_context");
    d(this, "_win");
    d(this, "_doc");
    d(this, "_scopeEvents");
    this._actions = n.actions, this.target = t, this._context = n.context || o, this._win = gt(bn(t) ? this._context : t), this._doc = this._win.document, this._scopeEvents = i, this.set(n);
  }
  /** @internal */
  get _defaults() {
    return {
      base: {},
      perAction: {},
      actions: {}
    };
  }
  setOnEvents(t, n) {
    return g.func(n.onstart) && this.on(`${t}start`, n.onstart), g.func(n.onmove) && this.on(`${t}move`, n.onmove), g.func(n.onend) && this.on(`${t}end`, n.onend), g.func(n.oninertiastart) && this.on(`${t}inertiastart`, n.oninertiastart), this;
  }
  updatePerActionListeners(t, n, o) {
    (g.array(n) || g.object(n)) && this.off(t, n), (g.array(o) || g.object(o)) && this.on(t, o);
  }
  setPerAction(t, n) {
    const o = this._defaults;
    for (const i in n) {
      const r = i, s = this.options[t], a = n[r];
      r === "listeners" && this.updatePerActionListeners(t, s.listeners, a), g.array(a) ? s[r] = nn(a) : g.plainObject(a) ? (s[r] = P(s[r] || {}, Ht(a)), g.object(o.perAction[r]) && "enabled" in o.perAction[r] && (s[r].enabled = a.enabled !== !1)) : g.bool(a) && g.object(o.perAction[r]) ? s[r].enabled = a : s[r] = a;
    }
  }
  /**
   * The default function to get an Interactables bounding rect. Can be
   * overridden using {@link Interactable.rectChecker}.
   *
   * @param {Element} [element] The element to measure.
   * @return {Rect} The object's bounding rectangle.
   */
  getRect(t) {
    return t = t || (g.element(this.target) ? this.target : null), g.string(this.target) && (t = t || this._context.querySelector(this.target)), tn(t);
  }
  /**
   * Returns or sets the function used to calculate the interactable's
   * element's rectangle
   *
   * @param {function} [checker] A function which returns this Interactable's
   * bounding rectangle. See {@link Interactable.getRect}
   * @return {function | object} The checker function or this Interactable
   */
  rectChecker(t) {
    return g.func(t) ? (this.getRect = (n) => {
      const o = P({}, t.apply(this, n));
      return "width" in o || (o.width = o.right - o.left, o.height = o.bottom - o.top), o;
    }, this) : t === null ? (delete this.getRect, this) : this.getRect;
  }
  _backCompatOption(t, n) {
    if (bn(n) || g.object(n)) {
      this.options[t] = n;
      for (const o in this._actions.map)
        this.options[o][t] = n;
      return this;
    }
    return this.options[t];
  }
  /**
   * Gets or sets the origin of the Interactable's element.  The x and y
   * of the origin will be subtracted from action event coordinates.
   *
   * @param {Element | object | string} [origin] An HTML or SVG Element whose
   * rect will be used, an object eg. { x: 0, y: 0 } or string 'parent', 'self'
   * or any CSS selector
   *
   * @return {object} The current origin or this Interactable
   */
  origin(t) {
    return this._backCompatOption("origin", t);
  }
  /**
   * Returns or sets the mouse coordinate types used to calculate the
   * movement of the pointer.
   *
   * @param {string} [newValue] Use 'client' if you will be scrolling while
   * interacting; Use 'page' if you want autoScroll to work
   * @return {string | object} The current deltaSource or this Interactable
   */
  deltaSource(t) {
    return t === "page" || t === "client" ? (this.options.deltaSource = t, this) : this.options.deltaSource;
  }
  /**
   * Gets the selector context Node of the Interactable. The default is
   * `window.document`.
   *
   * @return {Node} The context Node of this Interactable
   */
  context() {
    return this._context;
  }
  inContext(t) {
    return this._context === t.ownerDocument || Ot(this._context, t);
  }
  testIgnoreAllow(t, n, o) {
    return !this.testIgnore(t.ignoreFrom, n, o) && this.testAllow(t.allowFrom, n, o);
  }
  testAllow(t, n, o) {
    return t ? g.element(o) ? g.string(t) ? ke(o, t, n) : g.element(t) ? Ot(t, o) : !1 : !1 : !0;
  }
  testIgnore(t, n, o) {
    return !t || !g.element(o) ? !1 : g.string(t) ? ke(o, t, n) : g.element(t) ? Ot(t, o) : !1;
  }
  /**
   * Calls listeners for the given InteractEvent type bound globally
   * and directly to this Interactable
   *
   * @param {InteractEvent} iEvent The InteractEvent object to be fired on this
   * Interactable
   * @return {Interactable} this Interactable
   */
  fire(t) {
    return this.events.fire(t), this;
  }
  _onOff(t, n, o, i) {
    g.object(n) && !g.array(n) && (i = o, o = null);
    const r = t === "on" ? "add" : "remove", s = Ct(n, o);
    for (let a in s) {
      a === "wheel" && (a = rt.wheelEvent);
      for (const l of s[a])
        Ge(a, this._actions) ? this.events[t](a, l) : g.string(this.target) ? this._scopeEvents[`${r}Delegate`](this.target, this._context, a, l, i) : this._scopeEvents[r](this.target, a, l, i);
    }
    return this;
  }
  /**
   * Binds a listener for an InteractEvent, pointerEvent or DOM event.
   *
   * @param {string | array | object} types The types of events to listen
   * for
   * @param {function | array | object} [listener] The event listener function(s)
   * @param {object | boolean} [options] options object or useCapture flag for
   * addEventListener
   * @return {Interactable} This Interactable
   */
  on(t, n, o) {
    return this._onOff("on", t, n, o);
  }
  /**
   * Removes an InteractEvent, pointerEvent or DOM event listener.
   *
   * @param {string | array | object} types The types of events that were
   * listened for
   * @param {function | array | object} [listener] The event listener function(s)
   * @param {object | boolean} [options] options object or useCapture flag for
   * removeEventListener
   * @return {Interactable} This Interactable
   */
  off(t, n, o) {
    return this._onOff("off", t, n, o);
  }
  /**
   * Reset the options of this Interactable
   *
   * @param {object} options The new settings to apply
   * @return {object} This Interactable
   */
  set(t) {
    const n = this._defaults;
    g.object(t) || (t = {}), this.options = Ht(n.base);
    for (const o in this._actions.methodDict) {
      const i = o, r = this._actions.methodDict[i];
      this.options[i] = {}, this.setPerAction(i, P(P({}, n.perAction), n.actions[i])), this[r](t[i]);
    }
    for (const o in t)
      g.func(this[o]) && this[o](t[o]);
    return this;
  }
  /**
   * Remove this interactable from the list of interactables and remove it's
   * action capabilities and event listeners
   */
  unset() {
    if (g.string(this.target))
      for (const t in this._scopeEvents.delegatedEvents) {
        const n = this._scopeEvents.delegatedEvents[t];
        for (let o = n.length - 1; o >= 0; o--) {
          const {
            selector: i,
            context: r,
            listeners: s
          } = n[o];
          i === this.target && r === this._context && n.splice(o, 1);
          for (let a = s.length - 1; a >= 0; a--)
            this._scopeEvents.removeDelegate(this.target, this._context, t, s[a][0], s[a][1]);
        }
      }
    else
      this._scopeEvents.remove(this.target, "all");
  }
}
class is {
  constructor(t) {
    // all set interactables
    d(this, "list", []);
    d(this, "selectorMap", {});
    d(this, "scope");
    this.scope = t, t.addListeners({
      "interactable:unset": ({
        interactable: n
      }) => {
        const {
          target: o,
          _context: i
        } = n, r = g.string(o) ? this.selectorMap[o] : o[this.scope.id], s = ne(r, (a) => a.context === i);
        r[s] && (r[s].context = null, r[s].interactable = null), r.splice(s, 1);
      }
    });
  }
  new(t, n) {
    n = P(n || {}, {
      actions: this.scope.actions
    });
    const o = new this.scope.Interactable(t, n, this.scope.document, this.scope.events), i = {
      context: o._context,
      interactable: o
    };
    return this.scope.addDocument(o._doc), this.list.push(o), g.string(t) ? (this.selectorMap[t] || (this.selectorMap[t] = []), this.selectorMap[t].push(i)) : (o.target[this.scope.id] || Object.defineProperty(t, this.scope.id, {
      value: [],
      configurable: !0
    }), t[this.scope.id].push(i)), this.scope.fire("interactable:new", {
      target: t,
      options: n,
      interactable: o,
      win: this.scope._win
    }), o;
  }
  get(t, n) {
    const o = n && n.context || this.scope.document, i = g.string(t), r = i ? this.selectorMap[t] : t[this.scope.id];
    if (!r)
      return null;
    const s = ge(r, (a) => a.context === o && (i || a.interactable.inContext(t)));
    return s && s.interactable;
  }
  forEachMatch(t, n) {
    for (const o of this.list) {
      let i;
      if ((g.string(o.target) ? (
        // target is a selector and the element matches
        g.element(t) && wt(t, o.target)
      ) : (
        // target is the element
        t === o.target
      )) && // the element is in context
      o.inContext(t) && (i = n(o)), i !== void 0)
        return i;
    }
  }
}
function rs(e) {
  var t;
  const n = [], o = {}, i = [], r = {
    add: s,
    remove: a,
    addDelegate: l,
    removeDelegate: c,
    delegateListener: u,
    delegateUseCapture: f,
    delegatedEvents: o,
    documents: i,
    targets: n,
    supportsOptions: !1,
    supportsPassive: !1
  };
  (t = e.document) == null || t.createElement("div").addEventListener("test", null, {
    get capture() {
      return r.supportsOptions = !0;
    },
    get passive() {
      return r.supportsPassive = !0;
    }
  }), e.events = r;
  function s(h, m, v, D) {
    const I = Bt(D);
    let R = ge(n, (_) => _.eventTarget === h);
    R || (R = {
      eventTarget: h,
      events: {}
    }, n.push(R)), R.events[m] || (R.events[m] = []), h.addEventListener && !Li(R.events[m], v) && (h.addEventListener(m, v, r.supportsOptions ? I : I.capture), R.events[m].push(v));
  }
  function a(h, m, v, D) {
    const I = Bt(D), R = ne(n, (p) => p.eventTarget === h), _ = n[R];
    if (!_ || !_.events)
      return;
    if (m === "all") {
      for (m in _.events)
        _.events.hasOwnProperty(m) && a(h, m, "all");
      return;
    }
    let C = !1;
    const y = _.events[m];
    if (y) {
      if (v === "all") {
        for (let p = y.length - 1; p >= 0; p--)
          a(h, m, y[p], I);
        return;
      } else
        for (let p = 0; p < y.length; p++)
          if (y[p] === v) {
            h.removeEventListener(m, v, r.supportsOptions ? I : I.capture), y.splice(p, 1), y.length === 0 && (delete _.events[m], C = !0);
            break;
          }
    }
    C && !Object.keys(_.events).length && n.splice(R, 1);
  }
  function l(h, m, v, D, I) {
    const R = Bt(I);
    if (!o[v]) {
      o[v] = [];
      for (const y of i)
        s(y, v, u), s(y, v, f, !0);
    }
    const _ = o[v];
    let C = ge(_, (y) => y.selector === h && y.context === m);
    C || (C = {
      selector: h,
      context: m,
      listeners: []
    }, _.push(C)), C.listeners.push([D, R]);
  }
  function c(h, m, v, D, I) {
    const R = Bt(I), _ = o[v];
    let C = !1, y;
    if (_)
      for (y = _.length - 1; y >= 0; y--) {
        const p = _[y];
        if (p.selector === h && p.context === m) {
          const {
            listeners: M
          } = p;
          for (let x = M.length - 1; x >= 0; x--) {
            const [F, {
              capture: $,
              passive: k
            }] = M[x];
            if (F === D && $ === R.capture && k === R.passive) {
              M.splice(x, 1), M.length || (_.splice(y, 1), a(m, v, u), a(m, v, f, !0)), C = !0;
              break;
            }
          }
          if (C)
            break;
        }
      }
  }
  function u(h, m) {
    const v = Bt(m), D = new ss(h), I = o[h.type], [R] = ho(h);
    let _ = R;
    for (; g.element(_); ) {
      for (let C = 0; C < I.length; C++) {
        const y = I[C], {
          selector: p,
          context: M
        } = y;
        if (wt(_, p) && Ot(M, R) && Ot(M, _)) {
          const {
            listeners: x
          } = y;
          D.currentTarget = _;
          for (const [F, {
            capture: $,
            passive: k
          }] of x)
            $ === v.capture && k === v.passive && F(D);
        }
      }
      _ = at(_);
    }
  }
  function f(h) {
    return u.call(this, h, !0);
  }
  return r;
}
class ss {
  constructor(t) {
    d(this, "currentTarget");
    d(this, "originalEvent");
    d(this, "type");
    this.originalEvent = t, pe(this, t);
  }
  preventOriginalDefault() {
    this.originalEvent.preventDefault();
  }
  stopPropagation() {
    this.originalEvent.stopPropagation();
  }
  stopImmediatePropagation() {
    this.originalEvent.stopImmediatePropagation();
  }
}
function Bt(e) {
  if (!g.object(e))
    return {
      capture: !!e,
      passive: !1
    };
  const t = P({}, e);
  return t.capture = !!e.capture, t.passive = !!e.passive, t;
}
const as = {
  id: "events",
  install: rs
}, Ue = {
  methodOrder: ["simulationResume", "mouseOrPen", "hasPointer", "idle"],
  search(e) {
    for (const t of Ue.methodOrder) {
      const n = Ue[t](e);
      if (n)
        return n;
    }
    return null;
  },
  // try to resume simulation with a new pointer
  simulationResume({
    pointerType: e,
    eventType: t,
    eventTarget: n,
    scope: o
  }) {
    if (!/down|start/i.test(t))
      return null;
    for (const i of o.interactions.list) {
      let r = n;
      if (i.simulation && i.simulation.allowResume && i.pointerType === e)
        for (; r; ) {
          if (r === i.element)
            return i;
          r = at(r);
        }
    }
    return null;
  },
  // if it's a mouse or pen interaction
  mouseOrPen({
    pointerId: e,
    pointerType: t,
    eventType: n,
    scope: o
  }) {
    if (t !== "mouse" && t !== "pen")
      return null;
    let i;
    for (const r of o.interactions.list)
      if (r.pointerType === t) {
        if (r.simulation && !Rn(r, e))
          continue;
        if (r.interacting())
          return r;
        i || (i = r);
      }
    if (i)
      return i;
    for (const r of o.interactions.list)
      if (r.pointerType === t && !(/down/i.test(n) && r.simulation))
        return r;
    return null;
  },
  // get interaction that has this pointer
  hasPointer({
    pointerId: e,
    scope: t
  }) {
    for (const n of t.interactions.list)
      if (Rn(n, e))
        return n;
    return null;
  },
  // get first idle interaction with a matching pointerType
  idle({
    pointerType: e,
    scope: t
  }) {
    for (const n of t.interactions.list) {
      if (n.pointers.length === 1) {
        const o = n.interactable;
        if (o && !(o.options.gesture && o.options.gesture.enabled))
          continue;
      } else if (n.pointers.length >= 2)
        continue;
      if (!n.interacting() && e === n.pointerType)
        return n;
    }
    return null;
  }
};
function Rn(e, t) {
  return e.pointers.some(({
    id: n
  }) => n === t);
}
const cs = Ue, Co = ["pointerDown", "pointerMove", "pointerUp", "updatePointer", "removePointer", "windowBlur"];
function ls(e) {
  const t = {};
  for (const r of Co)
    t[r] = Mo(r, e);
  const n = rt.pEventTypes;
  let o;
  Q.PointerEvent ? o = [{
    type: n.down,
    listener: i
  }, {
    type: n.down,
    listener: t.pointerDown
  }, {
    type: n.move,
    listener: t.pointerMove
  }, {
    type: n.up,
    listener: t.pointerUp
  }, {
    type: n.cancel,
    listener: t.pointerUp
  }] : o = [{
    type: "mousedown",
    listener: t.pointerDown
  }, {
    type: "mousemove",
    listener: t.pointerMove
  }, {
    type: "mouseup",
    listener: t.pointerUp
  }, {
    type: "touchstart",
    listener: i
  }, {
    type: "touchstart",
    listener: t.pointerDown
  }, {
    type: "touchmove",
    listener: t.pointerMove
  }, {
    type: "touchend",
    listener: t.pointerUp
  }, {
    type: "touchcancel",
    listener: t.pointerUp
  }], o.push({
    type: "blur",
    listener(r) {
      for (const s of e.interactions.list)
        s.documentBlur(r);
    }
  }), e.prevTouchTime = 0, e.Interaction = class extends jr {
    get pointerMoveTolerance() {
      return e.interactions.pointerMoveTolerance;
    }
    set pointerMoveTolerance(r) {
      e.interactions.pointerMoveTolerance = r;
    }
    _now() {
      return e.now();
    }
  }, e.interactions = {
    // all active and idle interactions
    list: [],
    new(r) {
      r.scopeFire = (a, l) => e.fire(a, l);
      const s = new e.Interaction(r);
      return e.interactions.list.push(s), s;
    },
    listeners: t,
    docEvents: o,
    pointerMoveTolerance: 1
  };
  function i() {
    for (const r of e.interactions.list)
      if (!(!r.pointerIsDown || r.pointerType !== "touch" || r._interacting))
        for (const s of r.pointers)
          e.documents.some(({
            doc: a
          }) => Ot(a, s.downTarget)) || r.removePointer(s.pointer, s.event);
  }
  e.usePlugin(Io);
}
function Mo(e, t) {
  return function(n) {
    const o = t.interactions.list, i = po(n), [r, s] = ho(n), a = [];
    if (/^touch/.test(n.type)) {
      t.prevTouchTime = t.now();
      for (const l of n.changedTouches) {
        const c = l, u = Qt(c), f = {
          pointer: c,
          pointerId: u,
          pointerType: i,
          eventType: n.type,
          eventTarget: r,
          curEventTarget: s,
          scope: t
        }, h = kn(f);
        a.push([f.pointer, f.eventTarget, f.curEventTarget, h]);
      }
    } else {
      let l = !1;
      if (!rt.supportsPointerEvent && /mouse/.test(n.type)) {
        for (let c = 0; c < o.length && !l; c++)
          l = o[c].pointerType !== "mouse" && o[c].pointerIsDown;
        l = l || t.now() - t.prevTouchTime < 500 || // on iOS and Firefox Mobile, MouseEvent.timeStamp is zero if simulated
        n.timeStamp === 0;
      }
      if (!l) {
        const c = {
          pointer: n,
          pointerId: Qt(n),
          pointerType: i,
          eventType: n.type,
          curEventTarget: s,
          eventTarget: r,
          scope: t
        }, u = kn(c);
        a.push([c.pointer, c.eventTarget, c.curEventTarget, u]);
      }
    }
    for (const [l, c, u, f] of a)
      f[e](l, n, c, u);
  };
}
function kn(e) {
  const {
    pointerType: t,
    scope: n
  } = e, i = {
    interaction: cs.search(e),
    searchDetails: e
  };
  return n.fire("interactions:find", i), i.interaction || n.interactions.new({
    pointerType: t
  });
}
function De({
  doc: e,
  scope: t,
  options: n
}, o) {
  const {
    interactions: {
      docEvents: i
    },
    events: r
  } = t, s = r[o];
  t.browser.isIOS && !n.events && (n.events = {
    passive: !1
  });
  for (const l in r.delegatedEvents)
    s(e, l, r.delegateListener), s(e, l, r.delegateUseCapture, !0);
  const a = n && n.events;
  for (const {
    type: l,
    listener: c
  } of i)
    s(e, l, c, a);
}
const us = {
  id: "core/interactions",
  install: ls,
  listeners: {
    "scope:add-document": (e) => De(e, "add"),
    "scope:remove-document": (e) => De(e, "remove"),
    "interactable:unset": ({
      interactable: e
    }, t) => {
      for (let n = t.interactions.list.length - 1; n >= 0; n--) {
        const o = t.interactions.list[n];
        o.interactable === e && (o.stop(), t.fire("interactions:destroy", {
          interaction: o
        }), o.destroy(), t.interactions.list.length > 2 && t.interactions.list.splice(n, 1));
      }
    }
  },
  onDocSignal: De,
  doOnInteractions: Mo,
  methodNames: Co
}, fs = us;
class ds {
  constructor() {
    d(this, "id", `__interact_scope_${Math.floor(Math.random() * 100)}`);
    d(this, "isInitialized", !1);
    d(this, "listenerMaps", []);
    d(this, "browser", rt);
    d(this, "defaults", Ht(Po));
    d(this, "Eventable", Oo);
    d(this, "actions", {
      map: {},
      phases: {
        start: !0,
        move: !0,
        end: !0
      },
      methodDict: {},
      phaselessTypes: {}
    });
    d(this, "interactStatic", ns(this));
    d(this, "InteractEvent", an);
    d(this, "Interactable");
    d(this, "interactables", new is(this));
    // main window
    d(this, "_win");
    // main document
    d(this, "document");
    // main window
    d(this, "window");
    // all documents being listened to
    d(this, "documents", []);
    d(this, "_plugins", {
      list: [],
      map: {}
    });
    d(this, "onWindowUnload", (t) => this.removeDocument(t.target));
    const t = this;
    this.Interactable = class extends os {
      get _defaults() {
        return t.defaults;
      }
      set(n) {
        return super.set(n), t.fire("interactable:set", {
          options: n,
          interactable: this
        }), this;
      }
      unset() {
        super.unset();
        const n = t.interactables.list.indexOf(this);
        n < 0 || (t.interactables.list.splice(n, 1), t.fire("interactable:unset", {
          interactable: this
        }));
      }
    };
  }
  addListeners(t, n) {
    this.listenerMaps.push({
      id: n,
      map: t
    });
  }
  fire(t, n) {
    for (const {
      map: {
        [t]: o
      }
    } of this.listenerMaps)
      if (o && o(n, this, t) === !1)
        return !1;
  }
  init(t) {
    return this.isInitialized ? this : ps(this, t);
  }
  pluginIsInstalled(t) {
    const {
      id: n
    } = t;
    return n ? !!this._plugins.map[n] : this._plugins.list.indexOf(t) !== -1;
  }
  usePlugin(t, n) {
    if (!this.isInitialized)
      return this;
    if (this.pluginIsInstalled(t))
      return this;
    if (t.id && (this._plugins.map[t.id] = t), this._plugins.list.push(t), t.install && t.install(this, n), t.listeners && t.before) {
      let o = 0;
      const i = this.listenerMaps.length, r = t.before.reduce((s, a) => (s[a] = !0, s[An(a)] = !0, s), {});
      for (; o < i; o++) {
        const s = this.listenerMaps[o].id;
        if (s && (r[s] || r[An(s)]))
          break;
      }
      this.listenerMaps.splice(o, 0, {
        id: t.id,
        map: t.listeners
      });
    } else
      t.listeners && this.listenerMaps.push({
        id: t.id,
        map: t.listeners
      });
    return this;
  }
  addDocument(t, n) {
    if (this.getDocIndex(t) !== -1)
      return !1;
    const o = gt(t);
    n = n ? P({}, n) : {}, this.documents.push({
      doc: t,
      options: n
    }), this.events.documents.push(t), t !== this.document && this.events.add(o, "unload", this.onWindowUnload), this.fire("scope:add-document", {
      doc: t,
      window: o,
      scope: this,
      options: n
    });
  }
  removeDocument(t) {
    const n = this.getDocIndex(t), o = gt(t), i = this.documents[n].options;
    this.events.remove(o, "unload", this.onWindowUnload), this.documents.splice(n, 1), this.events.documents.splice(n, 1), this.fire("scope:remove-document", {
      doc: t,
      window: o,
      scope: this,
      options: i
    });
  }
  getDocIndex(t) {
    for (let n = 0; n < this.documents.length; n++)
      if (this.documents[n].doc === t)
        return n;
    return -1;
  }
  getDocOptions(t) {
    const n = this.getDocIndex(t);
    return n === -1 ? null : this.documents[n].options;
  }
  now() {
    return (this.window.Date || Date).now();
  }
}
function ps(e, t) {
  return e.isInitialized = !0, g.window(t) && no(t), Q.init(t), rt.init(t), bt.init(t), e.window = t, e.document = t.document, e.usePlugin(fs), e.usePlugin(as), e;
}
function An(e) {
  return e && e.replace(/\/.*$/, "");
}
const Ro = new ds(), hs = Ro.interactStatic, Z = hs, gs = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : globalThis;
Ro.init(gs);
const ms = () => {
}, vs = () => {
}, ys = (e) => {
  const t = [["x", "y"], ["left", "top"], ["right", "bottom"], ["width", "height"]].filter(([o, i]) => o in e || i in e), n = (o, i) => {
    const {
      range: r,
      limits: s = {
        left: -1 / 0,
        right: 1 / 0,
        top: -1 / 0,
        bottom: 1 / 0
      },
      offset: a = {
        x: 0,
        y: 0
      }
    } = e, l = {
      range: r,
      grid: e,
      x: null,
      y: null
    };
    for (const [c, u] of t) {
      const f = Math.round((o - a.x) / e[c]), h = Math.round((i - a.y) / e[u]);
      l[c] = Math.max(s.left, Math.min(s.right, f * e[c] + a.x)), l[u] = Math.max(s.top, Math.min(s.bottom, h * e[u] + a.y));
    }
    return l;
  };
  return n.grid = e, n.coordFields = t, n;
}, bs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  edgeTarget: ms,
  elements: vs,
  grid: ys
}, Symbol.toStringTag, { value: "Module" })), xs = {
  id: "snappers",
  install(e) {
    const {
      interactStatic: t
    } = e;
    t.snappers = P(t.snappers || {}, bs), t.createSnapGrid = t.snappers.grid;
  }
}, ws = xs, Ss = {
  start(e) {
    const {
      state: t,
      rect: n,
      edges: o,
      pageCoords: i
    } = e;
    let {
      ratio: r,
      enabled: s
    } = t.options;
    const {
      equalDelta: a,
      modifiers: l
    } = t.options;
    r === "preserve" && (r = n.width / n.height), t.startCoords = P({}, i), t.startRect = P({}, n), t.ratio = r, t.equalDelta = a;
    const c = t.linkedEdges = {
      top: o.top || o.left && !o.bottom,
      left: o.left || o.top && !o.right,
      bottom: o.bottom || o.right && !o.top,
      right: o.right || o.bottom && !o.left
    };
    if (t.xIsPrimaryAxis = !!(o.left || o.right), t.equalDelta) {
      const f = (c.left ? 1 : -1) * (c.top ? 1 : -1);
      t.edgeSign = {
        x: f,
        y: f
      };
    } else
      t.edgeSign = {
        x: c.left ? -1 : 1,
        y: c.top ? -1 : 1
      };
    if (s !== !1 && P(o, c), !(l != null && l.length))
      return;
    const u = new sn(e.interaction);
    u.copyFrom(e.interaction.modification), u.prepareStates(l), t.subModification = u, u.startAll({
      ...e
    });
  },
  set(e) {
    const {
      state: t,
      rect: n,
      coords: o
    } = e, {
      linkedEdges: i
    } = t, r = P({}, o), s = t.equalDelta ? Es : Is;
    if (P(e.edges, i), s(t, t.xIsPrimaryAxis, o, n), !t.subModification)
      return null;
    const a = P({}, n);
    we(i, a, {
      x: o.x - r.x,
      y: o.y - r.y
    });
    const l = t.subModification.setAll({
      ...e,
      rect: a,
      edges: i,
      pageCoords: o,
      prevCoords: o,
      prevRect: a
    }), {
      delta: c
    } = l;
    if (l.changed) {
      const u = Math.abs(c.x) > Math.abs(c.y);
      s(t, u, l.coords, l.rect), P(o, l.coords);
    }
    return l.eventProps;
  },
  defaults: {
    ratio: "preserve",
    equalDelta: !1,
    modifiers: [],
    enabled: !1
  }
};
function Es({
  startCoords: e,
  edgeSign: t
}, n, o) {
  n ? o.y = e.y + (o.x - e.x) * t.y : o.x = e.x + (o.y - e.y) * t.x;
}
function Is({
  startRect: e,
  startCoords: t,
  ratio: n,
  edgeSign: o
}, i, r, s) {
  if (i) {
    const a = s.width / n;
    r.y = t.y + (a - e.height) * o.y;
  } else {
    const a = s.height * n;
    r.x = t.x + (a - e.width) * o.x;
  }
}
const zs = Et(Ss, "aspectRatio"), ko = () => {
};
ko._defaults = {};
const ce = ko;
function Ts({
  rect: e,
  startOffset: t,
  state: n,
  interaction: o,
  pageCoords: i
}) {
  const {
    options: r
  } = n, {
    elementRect: s
  } = r, a = P({
    left: 0,
    top: 0,
    right: 0,
    bottom: 0
  }, r.offset || {});
  if (e && s) {
    const l = Mt(r.restriction, o, i);
    if (l) {
      const c = l.right - l.left - e.width, u = l.bottom - l.top - e.height;
      c < 0 && (a.left += c, a.right += c), u < 0 && (a.top += u, a.bottom += u);
    }
    a.left += t.left - e.width * s.left, a.top += t.top - e.height * s.top, a.right += t.right - e.width * (1 - s.right), a.bottom += t.bottom - e.height * (1 - s.bottom);
  }
  n.offset = a;
}
function Ps({
  coords: e,
  interaction: t,
  state: n
}) {
  const {
    options: o,
    offset: i
  } = n, r = Mt(o.restriction, t, e);
  if (!r)
    return;
  const s = Ci(r);
  e.x = Math.max(Math.min(s.right - i.right, e.x), s.left + i.left), e.y = Math.max(Math.min(s.bottom - i.bottom, e.y), s.top + i.top);
}
function Mt(e, t, n) {
  return g.func(e) ? Jt(e, t.interactable, t.element, [n.x, n.y, t]) : Jt(e, t.interactable, t.element);
}
const _s = {
  restriction: null,
  elementRect: null,
  offset: null,
  endOnly: !1,
  enabled: !1
}, ve = {
  start: Ts,
  set: Ps,
  defaults: _s
}, Ds = Et(ve, "restrict"), Ao = {
  top: 1 / 0,
  left: 1 / 0,
  bottom: -1 / 0,
  right: -1 / 0
}, No = {
  top: -1 / 0,
  left: -1 / 0,
  bottom: 1 / 0,
  right: 1 / 0
};
function Os({
  interaction: e,
  startOffset: t,
  state: n
}) {
  const {
    options: o
  } = n;
  let i;
  if (o) {
    const r = Mt(o.offset, e, e.coords.start.page);
    i = xe(r);
  }
  i = i || {
    x: 0,
    y: 0
  }, n.offset = {
    top: i.y + t.top,
    left: i.x + t.left,
    bottom: i.y - t.bottom,
    right: i.x - t.right
  };
}
function Cs({
  coords: e,
  edges: t,
  interaction: n,
  state: o
}) {
  const {
    offset: i,
    options: r
  } = o;
  if (!t)
    return;
  const s = P({}, e), a = Mt(r.inner, n, s) || {}, l = Mt(r.outer, n, s) || {};
  Nn(a, Ao), Nn(l, No), t.top ? e.y = Math.min(Math.max(l.top + i.top, s.y), a.top + i.top) : t.bottom && (e.y = Math.max(Math.min(l.bottom + i.bottom, s.y), a.bottom + i.bottom)), t.left ? e.x = Math.min(Math.max(l.left + i.left, s.x), a.left + i.left) : t.right && (e.x = Math.max(Math.min(l.right + i.right, s.x), a.right + i.right));
}
function Nn(e, t) {
  for (const n of ["top", "left", "bottom", "right"])
    n in e || (e[n] = t[n]);
  return e;
}
const Ms = {
  inner: null,
  outer: null,
  offset: null,
  endOnly: !1,
  enabled: !1
}, qt = {
  noInner: Ao,
  noOuter: No,
  start: Os,
  set: Cs,
  defaults: Ms
}, Rs = Et(qt, "restrictEdges"), ks = P({
  get elementRect() {
    return {
      top: 0,
      left: 0,
      bottom: 1,
      right: 1
    };
  },
  set elementRect(e) {
  }
}, ve.defaults), As = {
  start: ve.start,
  set: ve.set,
  defaults: ks
}, Ns = Et(As, "restrictRect"), $s = {
  width: -1 / 0,
  height: -1 / 0
}, Ls = {
  width: 1 / 0,
  height: 1 / 0
};
function Hs(e) {
  return qt.start(e);
}
function Fs(e) {
  const {
    interaction: t,
    state: n,
    rect: o,
    edges: i
  } = e, {
    options: r
  } = n;
  if (!i)
    return;
  const s = Ae(Mt(r.min, t, e.coords)) || $s, a = Ae(Mt(r.max, t, e.coords)) || Ls;
  n.options = {
    endOnly: r.endOnly,
    inner: P({}, qt.noInner),
    outer: P({}, qt.noOuter)
  }, i.top ? (n.options.inner.top = o.bottom - s.height, n.options.outer.top = o.bottom - a.height) : i.bottom && (n.options.inner.bottom = o.top + s.height, n.options.outer.bottom = o.top + a.height), i.left ? (n.options.inner.left = o.right - s.width, n.options.outer.left = o.right - a.width) : i.right && (n.options.inner.right = o.left + s.width, n.options.outer.right = o.left + a.width), qt.set(e), n.options = r;
}
const js = {
  min: null,
  max: null,
  endOnly: !1,
  enabled: !1
}, Bs = {
  start: Hs,
  set: Fs,
  defaults: js
}, Ws = Et(Bs, "restrictSize");
function Ys(e) {
  const {
    interaction: t,
    interactable: n,
    element: o,
    rect: i,
    state: r,
    startOffset: s
  } = e, {
    options: a
  } = r, l = a.offsetWithOrigin ? Gs(e) : {
    x: 0,
    y: 0
  };
  let c;
  if (a.offset === "startCoords")
    c = {
      x: t.coords.start.page.x,
      y: t.coords.start.page.y
    };
  else {
    const f = Jt(a.offset, n, o, [t]);
    c = xe(f) || {
      x: 0,
      y: 0
    }, c.x += l.x, c.y += l.y;
  }
  const {
    relativePoints: u
  } = a;
  r.offsets = i && u && u.length ? u.map((f, h) => ({
    index: h,
    relativePoint: f,
    x: s.left - i.width * f.x + c.x,
    y: s.top - i.height * f.y + c.y
  })) : [{
    index: 0,
    relativePoint: null,
    x: c.x,
    y: c.y
  }];
}
function Xs(e) {
  const {
    interaction: t,
    coords: n,
    state: o
  } = e, {
    options: i,
    offsets: r
  } = o, s = te(t.interactable, t.element, t.prepared.name), a = P({}, n), l = [];
  i.offsetWithOrigin || (a.x -= s.x, a.y -= s.y);
  for (const u of r) {
    const f = a.x - u.x, h = a.y - u.y;
    for (let m = 0, v = i.targets.length; m < v; m++) {
      const D = i.targets[m];
      let I;
      g.func(D) ? I = D(f, h, t._proxy, u, m) : I = D, I && l.push({
        x: (g.number(I.x) ? I.x : f) + u.x,
        y: (g.number(I.y) ? I.y : h) + u.y,
        range: g.number(I.range) ? I.range : i.range,
        source: D,
        index: m,
        offset: u
      });
    }
  }
  const c = {
    target: null,
    inRange: !1,
    distance: 0,
    range: 0,
    delta: {
      x: 0,
      y: 0
    }
  };
  for (const u of l) {
    const f = u.range, h = u.x - a.x, m = u.y - a.y, v = ee(h, m);
    let D = v <= f;
    f === 1 / 0 && c.inRange && c.range !== 1 / 0 && (D = !1), (!c.target || (D ? (
      // is the closest target in range?
      c.inRange && f !== 1 / 0 ? (
        // the pointer is relatively deeper in this target
        v / f < c.distance / c.range
      ) : (
        // this target has Infinite range and the closest doesn't
        f === 1 / 0 && c.range !== 1 / 0 || // OR this target is closer that the previous closest
        v < c.distance
      )
    ) : (
      // The other is not in range and the pointer is closer to this target
      !c.inRange && v < c.distance
    ))) && (c.target = u, c.distance = v, c.range = f, c.inRange = D, c.delta.x = h, c.delta.y = m);
  }
  return c.inRange && (n.x = c.target.x, n.y = c.target.y), o.closest = c, c;
}
function Gs(e) {
  const {
    element: t
  } = e.interaction;
  return xe(Jt(e.state.options.origin, null, null, [t])) || te(e.interactable, t, e.interaction.prepared.name);
}
const Us = {
  range: 1 / 0,
  targets: null,
  offset: null,
  offsetWithOrigin: !0,
  origin: null,
  relativePoints: null,
  endOnly: !1,
  enabled: !1
}, cn = {
  start: Ys,
  set: Xs,
  defaults: Us
}, qs = Et(cn, "snap");
function Vs(e) {
  const {
    state: t,
    edges: n
  } = e, {
    options: o
  } = t;
  if (!n)
    return null;
  e.state = {
    options: {
      targets: null,
      relativePoints: [{
        x: n.left ? 0 : 1,
        y: n.top ? 0 : 1
      }],
      offset: o.offset || "self",
      origin: {
        x: 0,
        y: 0
      },
      range: o.range
    }
  }, t.targetFields = t.targetFields || [["width", "height"], ["x", "y"]], cn.start(e), t.offsets = e.state.offsets, e.state = t;
}
function Ks(e) {
  const {
    interaction: t,
    state: n,
    coords: o
  } = e, {
    options: i,
    offsets: r
  } = n, s = {
    x: o.x - r[0].x,
    y: o.y - r[0].y
  };
  n.options = P({}, i), n.options.targets = [];
  for (const l of i.targets || []) {
    let c;
    if (g.func(l) ? c = l(s.x, s.y, t) : c = l, !!c) {
      for (const [u, f] of n.targetFields)
        if (u in c || f in c) {
          c.x = c[u], c.y = c[f];
          break;
        }
      n.options.targets.push(c);
    }
  }
  const a = cn.set(e);
  return n.options = i, a;
}
const Js = {
  range: 1 / 0,
  targets: null,
  offset: null,
  endOnly: !1,
  enabled: !1
}, ye = {
  start: Vs,
  set: Ks,
  defaults: Js
}, Qs = Et(ye, "snapSize");
function Zs(e) {
  const {
    edges: t
  } = e;
  return t ? (e.state.targetFields = e.state.targetFields || [[t.left ? "left" : "right", t.top ? "top" : "bottom"]], ye.start(e)) : null;
}
const ta = {
  start: Zs,
  set: ye.set,
  defaults: P(Ht(ye.defaults), {
    targets: void 0,
    range: void 0,
    offset: {
      x: 0,
      y: 0
    }
  })
}, ea = Et(ta, "snapEdges"), Oe = {
  aspectRatio: zs,
  restrictEdges: Rs,
  restrict: Ds,
  restrictRect: Ns,
  restrictSize: Ws,
  snapEdges: ea,
  snap: qs,
  snapSize: Qs,
  spring: ce,
  avoid: ce,
  transform: ce,
  rubberband: ce
}, na = {
  id: "modifiers",
  install(e) {
    const {
      interactStatic: t
    } = e;
    e.usePlugin(To), e.usePlugin(ws), t.modifiers = Oe;
    for (const n in Oe) {
      const {
        _defaults: o,
        _methods: i
      } = Oe[n];
      o._methods = i, e.defaults.perAction[n] = o;
    }
  }
}, oa = na;
class $o extends Se {
  /** */
  constructor(t, n, o, i, r, s) {
    if (super(r), pe(this, o), o !== n && pe(this, n), this.timeStamp = s, this.originalEvent = o, this.type = t, this.pointerId = Qt(n), this.pointerType = po(n), this.target = i, this.currentTarget = null, t === "tap") {
      const a = r.getPointerIndex(n);
      this.dt = this.timeStamp - r.pointers[a].downTime;
      const l = this.timeStamp - r.tapTime;
      this.double = !!r.prevTap && r.prevTap.type !== "doubletap" && r.prevTap.target === this.target && l < 500;
    } else
      t === "doubletap" && (this.dt = n.timeStamp - r.tapTime, this.double = !0);
  }
  _subtractOrigin({
    x: t,
    y: n
  }) {
    return this.pageX -= t, this.pageY -= n, this.clientX -= t, this.clientY -= n, this;
  }
  _addOrigin({
    x: t,
    y: n
  }) {
    return this.pageX += t, this.pageY += n, this.clientX += t, this.clientY += n, this;
  }
  /**
   * Prevent the default behaviour of the original Event
   */
  preventDefault() {
    this.originalEvent.preventDefault();
  }
}
const ia = {
  holdDuration: 600,
  ignoreFrom: null,
  allowFrom: null,
  origin: {
    x: 0,
    y: 0
  }
}, Vt = {
  id: "pointer-events/base",
  before: ["inertia", "modifiers", "auto-start", "actions"],
  install: ua,
  listeners: {
    "interactions:new": ra,
    "interactions:update-pointer": sa,
    "interactions:move": aa,
    "interactions:down": (e, t) => {
      ca(e, t), xt(e, t);
    },
    "interactions:up": (e, t) => {
      qe(e), xt(e, t), la(e, t);
    },
    "interactions:cancel": (e, t) => {
      qe(e), xt(e, t);
    }
  },
  PointerEvent: $o,
  fire: xt,
  collectEventTargets: Lo,
  defaults: ia,
  types: {
    down: !0,
    move: !0,
    up: !0,
    cancel: !0,
    tap: !0,
    doubletap: !0,
    hold: !0
  }
};
function xt(e, t) {
  const {
    interaction: n,
    pointer: o,
    event: i,
    eventTarget: r,
    type: s,
    targets: a = Lo(e, t)
  } = e, l = new $o(s, o, i, r, n, t.now());
  t.fire("pointerEvents:new", {
    pointerEvent: l
  });
  const c = {
    interaction: n,
    pointer: o,
    event: i,
    eventTarget: r,
    targets: a,
    type: s,
    pointerEvent: l
  };
  for (let u = 0; u < a.length; u++) {
    const f = a[u];
    for (const m in f.props || {})
      l[m] = f.props[m];
    const h = te(f.eventable, f.node);
    if (l._subtractOrigin(h), l.eventable = f.eventable, l.currentTarget = f.node, f.eventable.fire(l), l._addOrigin(h), l.immediatePropagationStopped || l.propagationStopped && u + 1 < a.length && a[u + 1].node !== l.currentTarget)
      break;
  }
  if (t.fire("pointerEvents:fired", c), s === "tap") {
    const u = l.double ? xt({
      interaction: n,
      pointer: o,
      event: i,
      eventTarget: r,
      type: "doubletap"
    }, t) : l;
    n.prevTap = u, n.tapTime = u.timeStamp;
  }
  return l;
}
function Lo({
  interaction: e,
  pointer: t,
  event: n,
  eventTarget: o,
  type: i
}, r) {
  const s = e.getPointerIndex(t), a = e.pointers[s];
  if (i === "tap" && (e.pointerWasMoved || // or if the pointerup target is different to the pointerdown target
  !(a && a.downTarget === o)))
    return [];
  const l = so(o), c = {
    interaction: e,
    pointer: t,
    event: n,
    eventTarget: o,
    type: i,
    path: l,
    targets: [],
    node: null
  };
  for (const u of l)
    c.node = u, r.fire("pointerEvents:collect-targets", c);
  return i === "hold" && (c.targets = c.targets.filter((u) => {
    var f, h;
    return u.eventable.options.holdDuration === ((f = e.pointers[s]) == null || (h = f.hold) == null ? void 0 : h.duration);
  })), c.targets;
}
function ra({
  interaction: e
}) {
  e.prevTap = null, e.tapTime = 0;
}
function sa({
  down: e,
  pointerInfo: t
}) {
  !e && t.hold || (t.hold = {
    duration: 1 / 0,
    timeout: null
  });
}
function qe({
  interaction: e,
  pointerIndex: t
}) {
  const n = e.pointers[t].hold;
  n && n.timeout && (clearTimeout(n.timeout), n.timeout = null);
}
function aa(e, t) {
  const {
    interaction: n,
    pointer: o,
    event: i,
    eventTarget: r,
    duplicate: s
  } = e;
  !s && (!n.pointerIsDown || n.pointerWasMoved) && (n.pointerIsDown && qe(e), xt({
    interaction: n,
    pointer: o,
    event: i,
    eventTarget: r,
    type: "move"
  }, t));
}
function ca({
  interaction: e,
  pointer: t,
  event: n,
  eventTarget: o,
  pointerIndex: i
}, r) {
  const s = e.pointers[i].hold, a = so(o), l = {
    interaction: e,
    pointer: t,
    event: n,
    eventTarget: o,
    type: "hold",
    targets: [],
    path: a,
    node: null
  };
  for (const u of a)
    l.node = u, r.fire("pointerEvents:collect-targets", l);
  if (!l.targets.length)
    return;
  let c = 1 / 0;
  for (const u of l.targets) {
    const f = u.eventable.options.holdDuration;
    f < c && (c = f);
  }
  s.duration = c, s.timeout = setTimeout(() => {
    xt({
      interaction: e,
      eventTarget: o,
      pointer: t,
      event: n,
      type: "hold"
    }, r);
  }, c);
}
function la({
  interaction: e,
  pointer: t,
  event: n,
  eventTarget: o
}, i) {
  e.pointerWasMoved || xt({
    interaction: e,
    eventTarget: o,
    pointer: t,
    event: n,
    type: "tap"
  }, i);
}
function ua(e) {
  e.pointerEvents = Vt, e.defaults.actions.pointerEvents = Vt.defaults, P(e.actions.phaselessTypes, Vt.types);
}
const fa = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  default: Vt
}, Symbol.toStringTag, { value: "Module" }));
function da(e) {
  e.usePlugin(Vt);
  const {
    pointerEvents: t
  } = e;
  t.defaults.holdRepeatInterval = 0, t.types.holdrepeat = e.actions.phaselessTypes.holdrepeat = !0;
}
function pa({
  pointerEvent: e
}) {
  e.type === "hold" && (e.count = (e.count || 0) + 1);
}
function ha({
  interaction: e,
  pointerEvent: t,
  eventTarget: n,
  targets: o
}, i) {
  if (t.type !== "hold" || !o.length)
    return;
  const r = o[0].eventable.options.holdRepeatInterval;
  r <= 0 || (e.holdIntervalHandle = setTimeout(() => {
    i.pointerEvents.fire({
      interaction: e,
      eventTarget: n,
      type: "hold",
      pointer: t,
      event: t
    }, i);
  }, r));
}
function ga({
  interaction: e
}) {
  e.holdIntervalHandle && (clearInterval(e.holdIntervalHandle), e.holdIntervalHandle = null);
}
const ma = {
  id: "pointer-events/holdRepeat",
  install: da,
  listeners: ["move", "up", "cancel", "endall"].reduce((e, t) => (e[`pointerEvents:${t}`] = ga, e), {
    "pointerEvents:new": pa,
    "pointerEvents:fired": ha
  })
}, va = ma;
function ya(e) {
  const {
    Interactable: t
  } = e;
  t.prototype.pointerEvents = ba;
  const n = t.prototype._backCompatOption;
  t.prototype._backCompatOption = function(o, i) {
    const r = n.call(this, o, i);
    return r === this && (this.events.options[o] = i), r;
  };
}
function ba(e) {
  return P(this.events.options, e), this;
}
const xa = {
  id: "pointer-events/interactableTargets",
  install: ya,
  listeners: {
    "pointerEvents:collect-targets": ({
      targets: e,
      node: t,
      type: n,
      eventTarget: o
    }, i) => {
      i.interactables.forEachMatch(t, (r) => {
        const s = r.events, a = s.options;
        s.types[n] && s.types[n].length && r.testIgnoreAllow(a, t, o) && e.push({
          node: t,
          eventable: s,
          props: {
            interactable: r
          }
        });
      });
    },
    "interactable:new": ({
      interactable: e
    }) => {
      e.events.getRect = function(t) {
        return e.getRect(t);
      };
    },
    "interactable:set": ({
      interactable: e,
      options: t
    }, n) => {
      P(e.events.options, n.pointerEvents.defaults), P(e.events.options, t.pointerEvents || {});
    }
  }
}, wa = xa, Sa = {
  id: "pointer-events",
  install(e) {
    e.usePlugin(fa), e.usePlugin(va), e.usePlugin(wa);
  }
}, Ea = Sa;
function Ia(e) {
  const {
    /** @lends Interactable */
    Interactable: t
  } = e;
  e.actions.phases.reflow = !0, t.prototype.reflow = function(n) {
    return za(this, n, e);
  };
}
function za(e, t, n) {
  const o = g.string(e.target) ? nn(e._context.querySelectorAll(e.target)) : [e.target], i = n.window.Promise, r = i ? [] : null;
  for (const s of o) {
    const a = e.getRect(s);
    if (!a)
      break;
    const l = ge(n.interactions.list, (u) => u.interacting() && u.interactable === e && u.element === s && u.prepared.name === t.name);
    let c;
    if (l)
      l.move(), r && (c = l._reflowPromise || new i((u) => {
        l._reflowResolve = u;
      }));
    else {
      const u = Ae(a), f = {
        page: {
          x: u.x,
          y: u.y
        },
        client: {
          x: u.x,
          y: u.y
        },
        timeStamp: n.now()
      }, h = $i(f);
      c = Ta(n, e, s, t, h);
    }
    r && r.push(c);
  }
  return r && i.all(r).then(() => e);
}
function Ta(e, t, n, o, i) {
  const r = e.interactions.new({
    pointerType: "reflow"
  }), s = {
    interaction: r,
    event: i,
    pointer: i,
    eventTarget: n,
    phase: "reflow"
  };
  r.interactable = t, r.element = n, r.prevEvent = i, r.updatePointer(i, i, n, !0), co(r.coords.delta), on(r.prepared, o), r._doPhase(s);
  const {
    Promise: a
  } = e.window, l = a ? new a((c) => {
    r._reflowResolve = c;
  }) : void 0;
  return r._reflowPromise = l, r.start(o, t, n), r._interacting ? (r.move(s), r.end(i)) : (r.stop(), r._reflowResolve()), r.removePointer(i, i), l;
}
const Pa = {
  id: "reflow",
  install: Ia,
  listeners: {
    // remove completed reflow interactions
    "interactions:stop": ({
      interaction: e
    }, t) => {
      e.pointerType === "reflow" && (e._reflowResolve && e._reflowResolve(), Hi(t.interactions.list, e));
    }
  }
}, _a = Pa;
Z.use(Io);
Z.use(Do);
Z.use(Ea);
Z.use(es);
Z.use(oa);
Z.use(zr);
Z.use(nr);
Z.use(sr);
Z.use(_a);
Z.use(Ar);
Z.default = Z;
const Ho = Symbol("GRID_PROVIDER"), Ve = Symbol("INTERSECTION_OBSERVER_ID"), $n = "vue-resizable-handle";
function Da() {
  const e = ii(Ho);
  if (e === void 0)
    throw new Error("You can't use useGridProvider outside of a GridProvider");
  return e;
}
const Oa = (e) => !Number.isNaN(e);
function Ca(e) {
  const t = e.target.offsetParent || document.body, n = e.offsetParent === document.body ? { left: 0, top: 0 } : t.getBoundingClientRect(), o = e.clientX + t.scrollLeft - n.left, i = e.clientY + t.scrollTop - n.top;
  return { x: o, y: i };
}
function Ln(e, t, n, o) {
  return !Oa(e) ? { deltaX: 0, deltaY: 0, lastX: n, lastY: o, x: n, y: o } : { deltaX: n - e, deltaY: o - t, lastX: e, lastY: t, x: n, y: o };
}
const Hn = (e) => Ca(e);
var _t = /* @__PURE__ */ ((e) => (e.DOWN = "DOWN", e.LEFT = "LEFT", e.RIGHT = "RIGHT", e.UP = "UP", e))(_t || {});
function Ma(e) {
  let t = 0, n;
  for (let o = 0; o < e.length; o++)
    n = e[o].y + e[o].h, n > t && (t = n);
  return t;
}
const Ra = (e) => JSON.parse(JSON.stringify(e));
function Ke(e) {
  const t = Array(e.length);
  for (let n = 0; n < e.length; n++)
    t[n] = Ra(e[n]);
  return t;
}
function Fo(e, t) {
  return !(e === t || e.x + e.w <= t.x || e.x >= t.x + t.w || e.y + e.h <= t.y || e.y >= t.y + t.h);
}
const jo = (e) => e.filter((t) => t.isStatic);
function Dt(e, t) {
  if (!e)
    return;
  const n = jo(e), o = Wo(e), i = Array(e.length);
  for (let r = 0; r < o.length; r++) {
    let s = o[r];
    s.isStatic || (s = ka(n, s, t), n.push(s)), i[e.indexOf(s)] = s, s.moved = !1;
  }
  return i;
}
function Kt(e, t) {
  for (let n = 0, o = e.length; n < o; n++)
    if (Fo(e[n], t))
      return e[n];
}
function ka(e, t, n) {
  if (n)
    for (; t.y > 0 && !Kt(e, t); )
      t.y--;
  let o = Kt(e, t);
  for (; o; )
    t.y = o.y + o.h, o = Kt(e, t);
  return t;
}
function Aa(e, t) {
  const n = jo(e);
  for (let o = 0; o < e.length; o++) {
    const i = e[o];
    if (i.x + i.w > t.cols && (i.x = t.cols - i.w), i.x < 0 && (i.x = 0, i.w = t.cols), !i.isStatic)
      n.push(i);
    else
      for (; Kt(n, i); )
        i.y++;
  }
  return e;
}
function Bo(e, t) {
  return e.filter((n) => Fo(n, t));
}
const Fn = (e, t) => e.filter((n) => n.id === t)[0];
function Je(e, t, n, o, i, r, s) {
  var h;
  if (t.isStatic)
    return e;
  const a = t.x, l = t.y, c = {
    DOWN: l < o,
    LEFT: a > n,
    RIGHT: a < n,
    UP: l > o
  };
  t.x = n, t.y = o, t.moved = !0;
  let u = Wo(e);
  c.UP && (u = u.reverse());
  const f = Bo(u, t);
  if (s && f.length)
    return t.x = a, t.y = l, t.moved = !1, e;
  for (let m = 0; m < f.length; m++) {
    const v = f[m];
    if (v.moved || t.y > v.y && t.y - v.y > v.h / 4)
      continue;
    const D = (h = Object.keys(c).filter((I) => c[I])) == null ? void 0 : h[0];
    v.isStatic ? e = jn(e, v, t, i, D, r) : e = jn(e, t, v, i, D, r);
  }
  return e;
}
function jn(e, t, n, o, i, r) {
  if (o) {
    const l = {
      h: n.h,
      id: -1,
      w: n.w,
      x: n.x,
      y: Math.max(t.y - n.h, 0)
    };
    if (!Kt(e, l))
      return Je(e, n, l.x, l.y, o, r, !1);
  }
  const a = {
    $default: {
      x: n.x,
      y: n.y + 1
    },
    [_t.LEFT]: [n.x + t.w, t.y],
    [_t.RIGHT]: [n.x - t.w, t.y],
    [_t.UP]: [n.x, n.y + t.h],
    [_t.DOWN]: [n.x, n.y - t.h]
  };
  if (r) {
    const l = i === _t.LEFT || i === _t.RIGHT;
    if (i in a && !(l && t.w < n.w && t.x !== n.x)) {
      const [c, u] = a[i];
      a.$default.x = c, a.$default.y = u;
    }
  }
  return Je(e, n, a.$default.x, a.$default.y, r, !1);
}
function Yt(e) {
  return e.length === 1 ? [e[0], e[0]] : e;
}
const Na = (e, t, n, o) => ({
  height: `${o}px`,
  left: `${t}px`,
  position: "absolute",
  top: `${e}px`,
  width: `${n}px`
});
function $a(e, t, n, o) {
  return {
    height: `${o}px`,
    position: "absolute",
    transform: `translate3d(${t}px,${e}px, 0)`,
    width: `${n}px`
  };
}
function Wo(e) {
  return [...e].sort((t, n) => t.y === n.y && t.x === n.x ? 0 : t.y > n.y || t.y === n.y && t.x > n.x ? 1 : -1);
}
function La(e, t, n) {
  return e.trim().replace(t, n);
}
const Ha = /* @__PURE__ */ Zn("div", { class: "corner-handle" }, null, -1), Fa = [
  Ha
], Qe = /* @__PURE__ */ Vn({
  __name: "GridItem",
  props: {
    breakpointCols: {},
    colNum: {},
    containerWidth: {},
    h: {},
    id: {},
    isDraggable: { type: Boolean },
    isResizable: { type: Boolean },
    lastBreakpoint: {},
    margin: {},
    maxRows: {},
    rowHeight: {},
    useCssTransforms: { type: Boolean },
    w: {},
    x: {},
    y: {},
    calculateStylesTrigger: {},
    dragAllowFrom: { default: null },
    dragIgnoreFrom: { default: "a, button" },
    dragOption: { default: () => ({}) },
    maxH: { default: Number.POSITIVE_INFINITY },
    maxW: { default: Number.POSITIVE_INFINITY },
    minH: { default: 1 },
    minW: { default: 1 },
    observer: { default: void 0 },
    isStatic: { type: Boolean, default: !1 }
  },
  emits: ["nocResizeContainer", "nocResize", "nocResizeEnd", "nocMove", "nocMoveEnd"],
  setup(e, { emit: t }) {
    const n = e, o = t, {
      handleDragEvent: i,
      handleResizeEvent: r
    } = Da(), s = X(null), a = X(!1), l = X({}), c = X({ h: n.h, w: n.w, x: n.x, y: n.y }), u = X(null), f = X(!1), h = X(!1), m = X({ h: Number.NaN, w: Number.NaN, x: Number.NaN, y: Number.NaN }), v = X({ h: Number.NaN, w: Number.NaN, x: Number.NaN, y: Number.NaN }), D = X(!1), I = X(null), R = ri({ props: {} }), _ = le(() => {
      const b = n.isDraggable || n.isResizable;
      return navigator.userAgent.toLowerCase().includes("android") && b && !n.isStatic;
    }), C = le(() => n.isResizable && !n.isStatic), y = le(() => ({
      "css-transforms": n.useCssTransforms,
      "disable-user-select": f.value,
      "no-touch": _.value,
      resizing: h.value,
      static: n.isStatic,
      "vue-draggable-dragging": f.value,
      "vue-resizable": C.value
    }));
    V(() => n.observer, () => {
      n.observer && s.value && (n.observer.observe(s.value), s.value[Ve] = n.id);
    }), V(() => n.colNum, () => {
      U(), k();
    }), V(() => n.containerWidth, () => {
      U(), k();
    }), V(() => n.h, (b) => {
      c.value.h = b, k();
    }), V(() => n.isDraggable, () => {
      H();
    }), V([
      () => n.isResizable,
      () => n.maxH,
      () => n.maxW,
      () => n.minH,
      () => n.minW
    ], () => {
      U();
    }), V(() => n.rowHeight, () => {
      k();
    }), V(() => n.isStatic, () => {
      U(), H();
    }), V(() => n.w, (b) => {
      c.value.w = b, $();
    }), V(() => n.x, (b) => {
      c.value.x = b, $();
    }), V(() => n.y, (b) => {
      c.value.y = b, $();
    }), V(() => n.calculateStylesTrigger, () => {
      $();
    });
    function p() {
      const [b] = Yt(n.margin);
      return (n.containerWidth - b * (n.colNum + 1)) / n.colNum;
    }
    function M({ h: b, w: L, x: w, y: T }) {
      const z = p(), S = ([q, B]) => ({
        height: b === Number.POSITIVE_INFINITY ? b : Math.round(n.rowHeight * b + Math.max(0, b - 1) * q),
        left: Math.floor(z * w + (w + 1) * B),
        top: Math.round(n.rowHeight * T + (T + 1) * q),
        width: L === Number.POSITIVE_INFINITY ? L : Math.round(z * L + Math.max(0, L - 1) * B)
      }), G = Yt(n.margin);
      return S(G);
    }
    function x(b, L) {
      const w = p(), [T, z] = Yt(n.margin), S = Math.round((L + T) / (w + T)), G = Math.round((b + z) / (n.rowHeight + z));
      return {
        h: Math.max(Math.min(G, n.maxRows - c.value.y), 0),
        w: Math.max(Math.min(S, n.colNum - c.value.x), 0)
      };
    }
    function F(b, L) {
      const w = p(), [T, z] = Yt(n.margin), S = Math.round((L - T) / (w + T)), G = Math.round((b - z) / (n.rowHeight + z));
      return {
        x: Math.max(Math.min(S, n.colNum - c.value.w), 0),
        y: Math.max(Math.min(G, n.maxRows - c.value.h), 0)
      };
    }
    function $() {
      var L, w;
      const b = M({
        h: c.value.h,
        w: c.value.w,
        x: c.value.x,
        y: c.value.y
      });
      n.x + n.w > n.colNum ? (c.value.x = 0, c.value.w = n.w > n.colNum ? n.colNum : n.w) : (c.value.x = n.x, c.value.w = n.w), f.value && (b.top = l.value.top ?? 0, b.left = l.value.left ?? 0), h.value && (b.width = ((L = I == null ? void 0 : I.value) == null ? void 0 : L.width) ?? 0, b.height = ((w = I == null ? void 0 : I.value) == null ? void 0 : w.height) ?? 0), R.props = n.useCssTransforms ? $a(b.top, b.left, b.width, b.height) : Na(b.top, b.left, b.width, b.height);
    }
    function k() {
      $();
      const b = {};
      for (const L of ["width", "height"]) {
        const w = R.props[L], T = w == null ? void 0 : w.toString().match(/^(\d+)px$/);
        if (!T)
          return;
        b[L] = +T[1];
      }
      o("nocResizeContainer", {
        h: n.h,
        height: b.height,
        id: n.id,
        w: n.w,
        width: b.width
      });
    }
    function W(b) {
      var G, q;
      if (n.isStatic || h.value)
        return;
      const L = Hn(b);
      if (!L)
        return;
      const { x: w, y: T } = L, z = { left: 0, top: 0 };
      switch (b.type) {
        case "dragstart": {
          v.value.x = c.value.x, v.value.y = c.value.y;
          const B = b.target.offsetParent.getBoundingClientRect(), tt = b.target.getBoundingClientRect();
          z.left = tt.left - B.left, z.top = tt.top - B.top, l.value = z, f.value = !0;
          break;
        }
        case "dragend": {
          if (!f.value)
            return;
          const B = b.target.offsetParent.getBoundingClientRect(), tt = b.target.getBoundingClientRect();
          z.left = tt.left - B.left, z.top = tt.top - B.top, l.value = {}, f.value = !1;
          break;
        }
        case "dragmove": {
          const B = Ln(m.value.x, m.value.y, w, T);
          z.left = (((G = l == null ? void 0 : l.value) == null ? void 0 : G.left) ?? 0) + B.deltaX, z.top = (((q = l == null ? void 0 : l.value) == null ? void 0 : q.top) ?? 0) + B.deltaY, l.value = z;
          break;
        }
      }
      const S = F(z.top, z.left);
      m.value.x = w, m.value.y = T, (c.value.x !== S.x || c.value.y !== S.y) && o("nocMove", {
        id: n.id,
        x: S.x,
        y: S.y
      }), b.type === "dragend" && (v.value.x !== c.value.x || v.value.y !== c.value.y) && o("nocMoveEnd", {
        id: n.id,
        x: S.x,
        y: S.y
      }), i({
        callback: $,
        eventType: b.type,
        h: c.value.h,
        id: n.id,
        w: c.value.w,
        x: S.x,
        y: S.y
      });
    }
    function A(b) {
      var G, q;
      if (n.isStatic)
        return;
      const L = Hn(b);
      if (!L)
        return;
      const { x: w, y: T } = L, z = { height: 0, width: 0 };
      switch (b.type) {
        case "resizestart": {
          v.value.w = c.value.w, v.value.h = c.value.h;
          const { height: B, width: tt } = M({
            h: c.value.h,
            w: c.value.w,
            x: c.value.x,
            y: c.value.y
          });
          z.width = tt, z.height = B, I.value = z, h.value = !0;
          break;
        }
        case "resizemove": {
          const B = Ln(m.value.x, m.value.h, w, T);
          z.width = (((G = I == null ? void 0 : I.value) == null ? void 0 : G.width) ?? 0) + B.deltaX, z.height = (((q = I == null ? void 0 : I.value) == null ? void 0 : q.height) ?? 0) + B.deltaY, I.value = z, h.value = !0;
          break;
        }
        case "resizeend": {
          const { height: B, width: tt } = M({
            h: c.value.h,
            w: c.value.w,
            x: c.value.x,
            y: c.value.y
          });
          z.width = tt, z.height = B, I.value = null, h.value = !1;
          break;
        }
      }
      const S = x(z.height, z.width);
      S.w < n.minW && (S.w = n.minW), S.w > n.maxW && (S.w = n.maxW), S.h < n.minH && (S.h = n.minH), S.h > n.maxH && (S.h = n.maxH), S.h < 1 && (S.h = 1), S.w < 1 && (S.w = 1), m.value.x = w, m.value.h = T, (c.value.w !== S.w || c.value.h !== S.h) && o("nocResize", {
        h: S.h,
        height: z.height,
        id: n.id,
        w: S.w,
        width: z.width
      }), b.type === "resizeend" && (v.value.w !== c.value.w || v.value.h !== c.value.h) && o("nocResizeEnd", {
        h: S.h,
        height: z.height,
        id: n.id,
        w: S.w,
        width: z.width
      }), r({
        callback: $,
        eventType: b.type,
        h: S.h,
        id: n.id,
        w: S.w,
        x: c.value.x,
        y: c.value.y
      });
    }
    function H() {
      !u.value && s.value && (u.value = Z(s.value)), n.isDraggable && !n.isStatic ? (u.value.draggable({
        allowFrom: n.dragAllowFrom,
        ignoreFrom: n.dragIgnoreFrom,
        ...n.dragOption
      }), a.value || (a.value = !0, u.value.on("dragstart dragmove dragend", W))) : u.value.draggable({ enabled: !1 });
    }
    function U() {
      if (!u.value && s.value && (u.value = Z(s.value)), n.isResizable && !n.isStatic) {
        const b = `.${La($n, " ", ".")}`, L = M({
          h: n.maxH,
          w: n.maxW,
          x: 0,
          y: 0
        }), w = M({
          h: n.minH,
          w: n.minW,
          x: 0,
          y: 0
        }), T = {
          edges: { bottom: b, left: !1, right: b, top: !1 },
          ignoreFrom: "a, button",
          restrictSize: {
            max: { height: L.height, width: L.width },
            min: { height: w.height, width: w.width }
          }
        };
        u.value.resizable(T), D.value || (D.value = !0, u.value.on("resizestart resizemove resizeend", A));
      } else
        u.value.resizable({ enabled: !1 });
    }
    return Kn(() => {
      u.value && u.value.unset(), n.observer && s.value !== null && n.observer.unobserve(s.value);
    }), Jn(() => {
      U(), H(), $();
    }), (b, L) => (Gt(), de("div", {
      ref_key: "item",
      ref: s,
      class: hn(["vue-grid-item", y.value]),
      style: Qn(R.props)
    }, [
      Ce(b.$slots, "default"),
      C.value ? (Gt(), de("div", {
        key: 0,
        class: hn(Me($n))
      }, Fa, 2)) : si("", !0)
    ], 6));
  }
});
function ja(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var Yo = { exports: {} }, Ba = Yo.exports = {};
Ba.forEach = function(e, t) {
  for (var n = 0; n < e.length; n++) {
    var o = t(e[n]);
    if (o)
      return o;
  }
};
var Xo = Yo.exports, Wa = function(e) {
  var t = e.stateHandler.getState;
  function n(s) {
    var a = t(s);
    return a && !!a.isDetectable;
  }
  function o(s) {
    t(s).isDetectable = !0;
  }
  function i(s) {
    return !!t(s).busy;
  }
  function r(s, a) {
    t(s).busy = !!a;
  }
  return {
    isDetectable: n,
    markAsDetectable: o,
    isBusy: i,
    markBusy: r
  };
}, Ya = function(e) {
  var t = {};
  function n(s) {
    var a = e.get(s);
    return a === void 0 ? [] : t[a] || [];
  }
  function o(s, a) {
    var l = e.get(s);
    t[l] || (t[l] = []), t[l].push(a);
  }
  function i(s, a) {
    for (var l = n(s), c = 0, u = l.length; c < u; ++c)
      if (l[c] === a) {
        l.splice(c, 1);
        break;
      }
  }
  function r(s) {
    var a = n(s);
    a && (a.length = 0);
  }
  return {
    get: n,
    add: o,
    removeListener: i,
    removeAllListeners: r
  };
}, Xa = function() {
  var e = 1;
  function t() {
    return e++;
  }
  return {
    generate: t
  };
}, Ga = function(e) {
  var t = e.idGenerator, n = e.stateHandler.getState;
  function o(r) {
    var s = n(r);
    return s && s.id !== void 0 ? s.id : null;
  }
  function i(r) {
    var s = n(r);
    if (!s)
      throw new Error("setId required the element to have a resize detection state.");
    var a = t.generate();
    return s.id = a, a;
  }
  return {
    get: o,
    set: i
  };
}, Ua = function(e) {
  function t() {
  }
  var n = {
    log: t,
    warn: t,
    error: t
  };
  if (!e && window.console) {
    var o = function(i, r) {
      i[r] = function() {
        var a = console[r];
        if (a.apply)
          a.apply(console, arguments);
        else
          for (var l = 0; l < arguments.length; l++)
            a(arguments[l]);
      };
    };
    o(n, "log"), o(n, "warn"), o(n, "error");
  }
  return n;
}, Go = { exports: {} }, Uo = Go.exports = {};
Uo.isIE = function(e) {
  function t() {
    var o = navigator.userAgent.toLowerCase();
    return o.indexOf("msie") !== -1 || o.indexOf("trident") !== -1 || o.indexOf(" edge/") !== -1;
  }
  if (!t())
    return !1;
  if (!e)
    return !0;
  var n = function() {
    var o, i = 3, r = document.createElement("div"), s = r.getElementsByTagName("i");
    do
      r.innerHTML = "<!--[if gt IE " + ++i + "]><i></i><![endif]-->";
    while (s[0]);
    return i > 4 ? i : o;
  }();
  return e === n;
};
Uo.isLegacyOpera = function() {
  return !!window.opera;
};
var qo = Go.exports, Vo = { exports: {} }, qa = Vo.exports = {};
qa.getOption = Va;
function Va(e, t, n) {
  var o = e[t];
  return o == null && n !== void 0 ? n : o;
}
var Ka = Vo.exports, Bn = Ka, Ja = function(t) {
  t = t || {};
  var n = t.reporter, o = Bn.getOption(t, "async", !0), i = Bn.getOption(t, "auto", !0);
  i && !o && (n && n.warn("Invalid options combination. auto=true and async=false is invalid. Setting async=true."), o = !0);
  var r = Wn(), s, a = !1;
  function l(v, D) {
    !a && i && o && r.size() === 0 && f(), r.add(v, D);
  }
  function c() {
    for (a = !0; r.size(); ) {
      var v = r;
      r = Wn(), v.process();
    }
    a = !1;
  }
  function u(v) {
    a || (v === void 0 && (v = o), s && (h(s), s = null), v ? f() : c());
  }
  function f() {
    s = m(c);
  }
  function h(v) {
    var D = clearTimeout;
    return D(v);
  }
  function m(v) {
    var D = function(I) {
      return setTimeout(I, 0);
    };
    return D(v);
  }
  return {
    add: l,
    force: u
  };
};
function Wn() {
  var e = {}, t = 0, n = 0, o = 0;
  function i(a, l) {
    l || (l = a, a = 0), a > n ? n = a : a < o && (o = a), e[a] || (e[a] = []), e[a].push(l), t++;
  }
  function r() {
    for (var a = o; a <= n; a++)
      for (var l = e[a], c = 0; c < l.length; c++) {
        var u = l[c];
        u();
      }
  }
  function s() {
    return t;
  }
  return {
    add: i,
    process: r,
    size: s
  };
}
var ln = "_erd";
function Qa(e) {
  return e[ln] = {}, Ko(e);
}
function Ko(e) {
  return e[ln];
}
function Za(e) {
  delete e[ln];
}
var tc = {
  initState: Qa,
  getState: Ko,
  cleanState: Za
}, Wt = qo, ec = function(e) {
  e = e || {};
  var t = e.reporter, n = e.batchProcessor, o = e.stateHandler.getState;
  if (!t)
    throw new Error("Missing required dependency: reporter.");
  function i(c, u) {
    function f() {
      u(c);
    }
    if (Wt.isIE(8))
      o(c).object = {
        proxy: f
      }, c.attachEvent("onresize", f);
    else {
      var h = a(c);
      if (!h)
        throw new Error("Element is not detectable by this strategy.");
      h.contentDocument.defaultView.addEventListener("resize", f);
    }
  }
  function r(c) {
    var u = e.important ? " !important; " : "; ";
    return (c.join(u) + u).trim();
  }
  function s(c, u, f) {
    f || (f = u, u = c, c = null), c = c || {}, c.debug;
    function h(m, v) {
      var D = r(["display: block", "position: absolute", "top: 0", "left: 0", "width: 100%", "height: 100%", "border: none", "padding: 0", "margin: 0", "opacity: 0", "z-index: -1000", "pointer-events: none"]), I = !1, R = window.getComputedStyle(m), _ = m.offsetWidth, C = m.offsetHeight;
      o(m).startSize = {
        width: _,
        height: C
      };
      function y() {
        function p() {
          if (R.position === "static") {
            m.style.setProperty("position", "relative", c.important ? "important" : "");
            var F = function($, k, W, A) {
              function H(b) {
                return b.replace(/[^-\d\.]/g, "");
              }
              var U = W[A];
              U !== "auto" && H(U) !== "0" && ($.warn("An element that is positioned static has style." + A + "=" + U + " which is ignored due to the static positioning. The element will need to be positioned relative, so the style." + A + " will be set to 0. Element: ", k), k.style.setProperty(A, "0", c.important ? "important" : ""));
            };
            F(t, m, R, "top"), F(t, m, R, "right"), F(t, m, R, "bottom"), F(t, m, R, "left");
          }
        }
        function M() {
          I || p();
          function F(k, W) {
            if (!k.contentDocument) {
              var A = o(k);
              A.checkForObjectDocumentTimeoutId && window.clearTimeout(A.checkForObjectDocumentTimeoutId), A.checkForObjectDocumentTimeoutId = setTimeout(function() {
                A.checkForObjectDocumentTimeoutId = 0, F(k, W);
              }, 100);
              return;
            }
            W(k.contentDocument);
          }
          var $ = this;
          F($, function(W) {
            v(m);
          });
        }
        R.position !== "" && (p(), I = !0);
        var x = document.createElement("object");
        x.style.cssText = D, x.tabIndex = -1, x.type = "text/html", x.setAttribute("aria-hidden", "true"), x.onload = M, Wt.isIE() || (x.data = "about:blank"), o(m) && (m.appendChild(x), o(m).object = x, Wt.isIE() && (x.data = "about:blank"));
      }
      n ? n.add(y) : y();
    }
    Wt.isIE(8) ? f(u) : h(u, f);
  }
  function a(c) {
    return o(c).object;
  }
  function l(c) {
    if (o(c)) {
      var u = a(c);
      u && (Wt.isIE(8) ? c.detachEvent("onresize", u.proxy) : c.removeChild(u), o(c).checkForObjectDocumentTimeoutId && window.clearTimeout(o(c).checkForObjectDocumentTimeoutId), delete o(c).object);
    }
  }
  return {
    makeDetectable: s,
    addListener: i,
    uninstall: l
  };
}, nc = Xo.forEach, oc = function(e) {
  e = e || {};
  var t = e.reporter, n = e.batchProcessor, o = e.stateHandler.getState;
  e.stateHandler.hasState;
  var i = e.idHandler;
  if (!n)
    throw new Error("Missing required dependency: batchProcessor");
  if (!t)
    throw new Error("Missing required dependency: reporter.");
  var r = u(), s = "erd_scroll_detection_scrollbar_style", a = "erd_scroll_detection_container";
  function l(y) {
    f(y, s, a);
  }
  l(window.document);
  function c(y) {
    var p = e.important ? " !important; " : "; ";
    return (y.join(p) + p).trim();
  }
  function u() {
    var y = 500, p = 500, M = document.createElement("div");
    M.style.cssText = c(["position: absolute", "width: " + y * 2 + "px", "height: " + p * 2 + "px", "visibility: hidden", "margin: 0", "padding: 0"]);
    var x = document.createElement("div");
    x.style.cssText = c(["position: absolute", "width: " + y + "px", "height: " + p + "px", "overflow: scroll", "visibility: none", "top: " + -y * 3 + "px", "left: " + -p * 3 + "px", "visibility: hidden", "margin: 0", "padding: 0"]), x.appendChild(M), document.body.insertBefore(x, document.body.firstChild);
    var F = y - x.clientWidth, $ = p - x.clientHeight;
    return document.body.removeChild(x), {
      width: F,
      height: $
    };
  }
  function f(y, p, M) {
    function x(W, A) {
      A = A || function(U) {
        y.head.appendChild(U);
      };
      var H = y.createElement("style");
      return H.innerHTML = W, H.id = p, A(H), H;
    }
    if (!y.getElementById(p)) {
      var F = M + "_animation", $ = M + "_animation_active", k = `/* Created by the element-resize-detector library. */
`;
      k += "." + M + " > div::-webkit-scrollbar { " + c(["display: none"]) + ` }

`, k += "." + $ + " { " + c(["-webkit-animation-duration: 0.1s", "animation-duration: 0.1s", "-webkit-animation-name: " + F, "animation-name: " + F]) + ` }
`, k += "@-webkit-keyframes " + F + ` { 0% { opacity: 1; } 50% { opacity: 0; } 100% { opacity: 1; } }
`, k += "@keyframes " + F + " { 0% { opacity: 1; } 50% { opacity: 0; } 100% { opacity: 1; } }", x(k);
    }
  }
  function h(y) {
    y.className += " " + a + "_animation_active";
  }
  function m(y, p, M) {
    if (y.addEventListener)
      y.addEventListener(p, M);
    else if (y.attachEvent)
      y.attachEvent("on" + p, M);
    else
      return t.error("[scroll] Don't know how to add event listeners.");
  }
  function v(y, p, M) {
    if (y.removeEventListener)
      y.removeEventListener(p, M);
    else if (y.detachEvent)
      y.detachEvent("on" + p, M);
    else
      return t.error("[scroll] Don't know how to remove event listeners.");
  }
  function D(y) {
    return o(y).container.childNodes[0].childNodes[0].childNodes[0];
  }
  function I(y) {
    return o(y).container.childNodes[0].childNodes[0].childNodes[1];
  }
  function R(y, p) {
    var M = o(y).listeners;
    if (!M.push)
      throw new Error("Cannot add listener to an element that is not detectable.");
    o(y).listeners.push(p);
  }
  function _(y, p, M) {
    M || (M = p, p = y, y = null), y = y || {};
    function x() {
      if (y.debug) {
        var E = Array.prototype.slice.call(arguments);
        if (E.unshift(i.get(p), "Scroll: "), t.log.apply)
          t.log.apply(null, E);
        else
          for (var N = 0; N < E.length; N++)
            t.log(E[N]);
      }
    }
    function F(E) {
      function N(Y) {
        var st = Y.getRootNode && Y.getRootNode().contains(Y);
        return Y === Y.ownerDocument.body || Y.ownerDocument.body.contains(Y) || st;
      }
      return !N(E) || window.getComputedStyle(E) === null;
    }
    function $(E) {
      var N = o(E).container.childNodes[0], Y = window.getComputedStyle(N);
      return !Y.width || Y.width.indexOf("px") === -1;
    }
    function k() {
      var E = window.getComputedStyle(p), N = {};
      return N.position = E.position, N.width = p.offsetWidth, N.height = p.offsetHeight, N.top = E.top, N.right = E.right, N.bottom = E.bottom, N.left = E.left, N.widthCSS = E.width, N.heightCSS = E.height, N;
    }
    function W() {
      var E = k();
      o(p).startSize = {
        width: E.width,
        height: E.height
      }, x("Element start size", o(p).startSize);
    }
    function A() {
      o(p).listeners = [];
    }
    function H() {
      if (x("storeStyle invoked."), !o(p)) {
        x("Aborting because element has been uninstalled");
        return;
      }
      var E = k();
      o(p).style = E;
    }
    function U(E, N, Y) {
      o(E).lastWidth = N, o(E).lastHeight = Y;
    }
    function b(E) {
      return D(E).childNodes[0];
    }
    function L() {
      return 2 * r.width + 1;
    }
    function w() {
      return 2 * r.height + 1;
    }
    function T(E) {
      return E + 10 + L();
    }
    function z(E) {
      return E + 10 + w();
    }
    function S(E) {
      return E * 2 + L();
    }
    function G(E) {
      return E * 2 + w();
    }
    function q(E, N, Y) {
      var st = D(E), It = I(E), Rt = T(N), kt = z(Y), et = S(N), j = G(Y);
      st.scrollLeft = Rt, st.scrollTop = kt, It.scrollLeft = et, It.scrollTop = j;
    }
    function B() {
      var E = o(p).container;
      if (!E) {
        E = document.createElement("div"), E.className = a, E.style.cssText = c(["visibility: hidden", "display: inline", "width: 0px", "height: 0px", "z-index: -1", "overflow: hidden", "margin: 0", "padding: 0"]), o(p).container = E, h(E), p.appendChild(E);
        var N = function() {
          o(p).onRendered && o(p).onRendered();
        };
        m(E, "animationstart", N), o(p).onAnimationStart = N;
      }
      return E;
    }
    function tt() {
      function E() {
        var J = o(p).style;
        if (J.position === "static") {
          p.style.setProperty("position", "relative", y.important ? "important" : "");
          var ft = function(Nt, Tt, Zo, re) {
            function ti(ei) {
              return ei.replace(/[^-\d\.]/g, "");
            }
            var Ie = Zo[re];
            Ie !== "auto" && ti(Ie) !== "0" && (Nt.warn("An element that is positioned static has style." + re + "=" + Ie + " which is ignored due to the static positioning. The element will need to be positioned relative, so the style." + re + " will be set to 0. Element: ", Tt), Tt.style[re] = 0);
          };
          ft(t, p, J, "top"), ft(t, p, J, "right"), ft(t, p, J, "bottom"), ft(t, p, J, "left");
        }
      }
      function N(J, ft, Nt, Tt) {
        return J = J ? J + "px" : "0", ft = ft ? ft + "px" : "0", Nt = Nt ? Nt + "px" : "0", Tt = Tt ? Tt + "px" : "0", ["left: " + J, "top: " + ft, "right: " + Tt, "bottom: " + Nt];
      }
      if (x("Injecting elements"), !o(p)) {
        x("Aborting because element has been uninstalled");
        return;
      }
      E();
      var Y = o(p).container;
      Y || (Y = B());
      var st = r.width, It = r.height, Rt = c(["position: absolute", "flex: none", "overflow: hidden", "z-index: -1", "visibility: hidden", "width: 100%", "height: 100%", "left: 0px", "top: 0px"]), kt = c(["position: absolute", "flex: none", "overflow: hidden", "z-index: -1", "visibility: hidden"].concat(N(-(1 + st), -(1 + It), -It, -st))), et = c(["position: absolute", "flex: none", "overflow: scroll", "z-index: -1", "visibility: hidden", "width: 100%", "height: 100%"]), j = c(["position: absolute", "flex: none", "overflow: scroll", "z-index: -1", "visibility: hidden", "width: 100%", "height: 100%"]), nt = c(["position: absolute", "left: 0", "top: 0"]), zt = c(["position: absolute", "width: 200%", "height: 200%"]), lt = document.createElement("div"), ut = document.createElement("div"), At = document.createElement("div"), un = document.createElement("div"), ie = document.createElement("div"), fn = document.createElement("div");
      lt.dir = "ltr", lt.style.cssText = Rt, lt.className = a, ut.className = a, ut.style.cssText = kt, At.style.cssText = et, un.style.cssText = nt, ie.style.cssText = j, fn.style.cssText = zt, At.appendChild(un), ie.appendChild(fn), ut.appendChild(At), ut.appendChild(ie), lt.appendChild(ut), Y.appendChild(lt);
      function dn() {
        var J = o(p);
        J && J.onExpand ? J.onExpand() : x("Aborting expand scroll handler: element has been uninstalled");
      }
      function pn() {
        var J = o(p);
        J && J.onShrink ? J.onShrink() : x("Aborting shrink scroll handler: element has been uninstalled");
      }
      m(At, "scroll", dn), m(ie, "scroll", pn), o(p).onExpandScroll = dn, o(p).onShrinkScroll = pn;
    }
    function K() {
      function E(et, j, nt) {
        var zt = b(et), lt = T(j), ut = z(nt);
        zt.style.setProperty("width", lt + "px", y.important ? "important" : ""), zt.style.setProperty("height", ut + "px", y.important ? "important" : "");
      }
      function N(et) {
        var j = p.offsetWidth, nt = p.offsetHeight, zt = j !== o(p).lastWidth || nt !== o(p).lastHeight;
        x("Storing current size", j, nt), U(p, j, nt), n.add(0, function() {
          if (zt) {
            if (!o(p)) {
              x("Aborting because element has been uninstalled");
              return;
            }
            if (!Y()) {
              x("Aborting because element container has not been initialized");
              return;
            }
            if (y.debug) {
              var ut = p.offsetWidth, At = p.offsetHeight;
              (ut !== j || At !== nt) && t.warn(i.get(p), "Scroll: Size changed before updating detector elements.");
            }
            E(p, j, nt);
          }
        }), n.add(1, function() {
          if (!o(p)) {
            x("Aborting because element has been uninstalled");
            return;
          }
          if (!Y()) {
            x("Aborting because element container has not been initialized");
            return;
          }
          q(p, j, nt);
        }), zt && et && n.add(2, function() {
          if (!o(p)) {
            x("Aborting because element has been uninstalled");
            return;
          }
          if (!Y()) {
            x("Aborting because element container has not been initialized");
            return;
          }
          et();
        });
      }
      function Y() {
        return !!o(p).container;
      }
      function st() {
        function et() {
          return o(p).lastNotifiedWidth === void 0;
        }
        x("notifyListenersIfNeeded invoked");
        var j = o(p);
        if (et() && j.lastWidth === j.startSize.width && j.lastHeight === j.startSize.height)
          return x("Not notifying: Size is the same as the start size, and there has been no notification yet.");
        if (j.lastWidth === j.lastNotifiedWidth && j.lastHeight === j.lastNotifiedHeight)
          return x("Not notifying: Size already notified");
        x("Current size not notified, notifying..."), j.lastNotifiedWidth = j.lastWidth, j.lastNotifiedHeight = j.lastHeight, nc(o(p).listeners, function(nt) {
          nt(p);
        });
      }
      function It() {
        if (x("startanimation triggered."), $(p)) {
          x("Ignoring since element is still unrendered...");
          return;
        }
        x("Element rendered.");
        var et = D(p), j = I(p);
        (et.scrollLeft === 0 || et.scrollTop === 0 || j.scrollLeft === 0 || j.scrollTop === 0) && (x("Scrollbars out of sync. Updating detector elements..."), N(st));
      }
      function Rt() {
        if (x("Scroll detected."), $(p)) {
          x("Scroll event fired while unrendered. Ignoring...");
          return;
        }
        N(st);
      }
      if (x("registerListenersAndPositionElements invoked."), !o(p)) {
        x("Aborting because element has been uninstalled");
        return;
      }
      o(p).onRendered = It, o(p).onExpand = Rt, o(p).onShrink = Rt;
      var kt = o(p).style;
      E(p, kt.width, kt.height);
    }
    function Ft() {
      if (x("finalizeDomMutation invoked."), !o(p)) {
        x("Aborting because element has been uninstalled");
        return;
      }
      var E = o(p).style;
      U(p, E.width, E.height), q(p, E.width, E.height);
    }
    function oe() {
      M(p);
    }
    function ct() {
      x("Installing..."), A(), W(), n.add(0, H), n.add(1, tt), n.add(2, K), n.add(3, Ft), n.add(4, oe);
    }
    x("Making detectable..."), F(p) ? (x("Element is detached"), B(), x("Waiting until element is attached..."), o(p).onRendered = function() {
      x("Element is now attached"), ct();
    }) : ct();
  }
  function C(y) {
    var p = o(y);
    p && (p.onExpandScroll && v(D(y), "scroll", p.onExpandScroll), p.onShrinkScroll && v(I(y), "scroll", p.onShrinkScroll), p.onAnimationStart && v(p.container, "animationstart", p.onAnimationStart), p.container && y.removeChild(p.container));
  }
  return {
    makeDetectable: _,
    addListener: R,
    uninstall: C,
    initDocument: l
  };
}, Xt = Xo.forEach, ic = Wa, rc = Ya, sc = Xa, ac = Ga, cc = Ua, Yn = qo, lc = Ja, vt = tc, uc = ec, fc = oc;
function Xn(e) {
  return Array.isArray(e) || e.length !== void 0;
}
function Gn(e) {
  if (Array.isArray(e))
    return e;
  var t = [];
  return Xt(e, function(n) {
    t.push(n);
  }), t;
}
function Un(e) {
  return e && e.nodeType === 1;
}
var dc = function(e) {
  e = e || {};
  var t;
  if (e.idHandler)
    t = {
      get: function(_) {
        return e.idHandler.get(_, !0);
      },
      set: e.idHandler.set
    };
  else {
    var n = sc(), o = ac({
      idGenerator: n,
      stateHandler: vt
    });
    t = o;
  }
  var i = e.reporter;
  if (!i) {
    var r = i === !1;
    i = cc(r);
  }
  var s = yt(e, "batchProcessor", lc({ reporter: i })), a = {};
  a.callOnAdd = !!yt(e, "callOnAdd", !0), a.debug = !!yt(e, "debug", !1);
  var l = rc(t), c = ic({
    stateHandler: vt
  }), u, f = yt(e, "strategy", "object"), h = yt(e, "important", !1), m = {
    reporter: i,
    batchProcessor: s,
    stateHandler: vt,
    idHandler: t,
    important: h
  };
  if (f === "scroll" && (Yn.isLegacyOpera() ? (i.warn("Scroll strategy is not supported on legacy Opera. Changing to object strategy."), f = "object") : Yn.isIE(9) && (i.warn("Scroll strategy is not supported on IE9. Changing to object strategy."), f = "object")), f === "scroll")
    u = fc(m);
  else if (f === "object")
    u = uc(m);
  else
    throw new Error("Invalid strategy name: " + f);
  var v = {};
  function D(_, C, y) {
    function p(W) {
      var A = l.get(W);
      Xt(A, function(U) {
        U(W);
      });
    }
    function M(W, A, H) {
      l.add(A, H), W && H(A);
    }
    if (y || (y = C, C = _, _ = {}), !C)
      throw new Error("At least one element required.");
    if (!y)
      throw new Error("Listener required.");
    if (Un(C))
      C = [C];
    else if (Xn(C))
      C = Gn(C);
    else
      return i.error("Invalid arguments. Must be a DOM element or a collection of DOM elements.");
    var x = 0, F = yt(_, "callOnAdd", a.callOnAdd), $ = yt(_, "onReady", function() {
    }), k = yt(_, "debug", a.debug);
    Xt(C, function(A) {
      vt.getState(A) || (vt.initState(A), t.set(A));
      var H = t.get(A);
      if (k && i.log("Attaching listener to element", H, A), !c.isDetectable(A)) {
        if (k && i.log(H, "Not detectable."), c.isBusy(A)) {
          k && i.log(H, "System busy making it detectable"), M(F, A, y), v[H] = v[H] || [], v[H].push(function() {
            x++, x === C.length && $();
          });
          return;
        }
        return k && i.log(H, "Making detectable..."), c.markBusy(A, !0), u.makeDetectable({ debug: k, important: h }, A, function(b) {
          if (k && i.log(H, "onElementDetectable"), vt.getState(b)) {
            c.markAsDetectable(b), c.markBusy(b, !1), u.addListener(b, p), M(F, b, y);
            var L = vt.getState(b);
            if (L && L.startSize) {
              var w = b.offsetWidth, T = b.offsetHeight;
              (L.startSize.width !== w || L.startSize.height !== T) && p(b);
            }
            v[H] && Xt(v[H], function(z) {
              z();
            });
          } else
            k && i.log(H, "Element uninstalled before being detectable.");
          delete v[H], x++, x === C.length && $();
        });
      }
      k && i.log(H, "Already detecable, adding listener."), M(F, A, y), x++;
    }), x === C.length && $();
  }
  function I(_) {
    if (!_)
      return i.error("At least one element is required.");
    if (Un(_))
      _ = [_];
    else if (Xn(_))
      _ = Gn(_);
    else
      return i.error("Invalid arguments. Must be a DOM element or a collection of DOM elements.");
    Xt(_, function(C) {
      l.removeAllListeners(C), u.uninstall(C), vt.cleanState(C);
    });
  }
  function R(_) {
    u.initDocument && u.initDocument(_);
  }
  return {
    listenTo: D,
    removeListener: l.removeListener,
    removeAllListeners: l.removeAllListeners,
    uninstall: I,
    initDocument: R
  };
};
function yt(e, t, n) {
  var o = e[t];
  return o == null && n !== void 0 ? n : o;
}
const pc = /* @__PURE__ */ ja(dc), Jo = () => typeof window < "u";
function hc(e, t) {
  if (!Jo)
    return t();
  window.addEventListener(e, t);
}
function gc(e, t) {
  Jo && window.removeEventListener(e, t);
}
const mc = (e, t, n, o, i, r, s) => {
  if (Object.prototype.hasOwnProperty.call(t, o))
    return Ke(t[o] || []);
  let a = e;
  const l = Qo(n), c = l.slice(l.indexOf(o));
  for (let u = 0; u < c.length; u++) {
    const f = c[u];
    if (Object.prototype.hasOwnProperty.call(t, f)) {
      a = t[f];
      break;
    }
  }
  return a = Ke(a || []), Dt(Aa(a, { cols: r }), s);
};
function vc(e, t) {
  const n = Qo(e);
  let [o] = n;
  for (let i = 1; i < n.length; i++) {
    const r = n[i];
    t > (e[r] ?? 1) && (o = r);
  }
  return o;
}
function qn(e, t) {
  return t[e];
}
function Qo(e) {
  return Object.keys(e).sort((t, n) => (e[t] ?? 1) - (e[n] ?? 1));
}
const yc = /* @__PURE__ */ Vn({
  __name: "GridLayout",
  props: {
    colNum: {},
    layout: {},
    autoSize: { type: Boolean, default: !0 },
    breakpoints: { default: () => ({ lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }) },
    cols: { default: () => ({ lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 }) },
    horizontalShift: { type: Boolean, default: !1 },
    intersectionObserverConfig: { default: () => ({ root: null, rootMargin: "8px", threshold: 0.4 }) },
    isDraggable: { type: Boolean, default: !0 },
    isResizable: { type: Boolean, default: !0 },
    margin: { default: () => [10] },
    maxRows: { default: Number.POSITIVE_INFINITY },
    preventCollision: { type: Boolean, default: !1 },
    responsive: { type: Boolean, default: !1 },
    responsiveLayouts: { default: () => ({}) },
    rowHeight: { default: 150 },
    useCssTransforms: { type: Boolean, default: !0 },
    useObserver: { type: Boolean, default: !1 },
    verticalCompact: { type: Boolean, default: !0 }
  },
  emits: ["nocResizeContainer", "nocLayoutBeforeMount", "nocLayoutCreate", "nocLayoutMount", "nocLayoutReady", "nocLayoutUpdate", "update:layout", "update:breakpoints", "nocIntersectionObserve", "nocIntersectionUnobserve", "nocItemMove", "nocItemMoveEnd", "nocItemResize", "nocItemResizeEnd"],
  setup(e, { emit: t }) {
    const n = e, o = t, i = [
      "minW",
      "minH",
      "maxW",
      "maxH",
      "moved",
      "static",
      "isDraggable",
      "isResizable",
      "isBounded",
      "dragIgnoreFrom",
      "dragAllowFrom",
      "resizeIgnoreFrom",
      "preserveAspectRatio",
      "dragOption",
      "resizeOption"
    ], r = { h: 0, id: -1, w: 0, x: 0, y: 0 }, s = X(pc({ callOnAdd: !1, strategy: "scroll" })), a = X(!1), l = X(""), c = X(0), u = X({}), f = X({}), h = X(n.layout), m = X({ h: 0, id: -1, w: 0, x: 0, y: 0 }), v = X(0);
    let D;
    const I = X(null), R = X(n.colNum), _ = X(!1), C = le(() => ({
      breakpointCols: n.cols,
      calculateStylesTrigger: _,
      colNum: R.value,
      containerWidth: v.value,
      isDraggable: n.isDraggable,
      isResizable: n.isResizable,
      lastBreakpoint: l.value,
      margin: n.margin,
      maxRows: n.maxRows,
      responsive: n.responsive,
      rowHeight: n.rowHeight,
      useCssTransforms: n.useCssTransforms,
      width: v.value
    }));
    V(() => n.colNum, (w) => {
      R.value = w;
    }), V(() => n.layout.length, () => {
      M(), Dt(n.layout, n.verticalCompact);
    }), V(() => n.margin, () => {
      $();
    }), V(() => n.layout, () => {
      $();
    }), V(() => n.responsive, (w) => {
      w || (o("update:layout", h.value), R.value = n.colNum), W();
    }), V(() => v.value, (w, T) => {
      mt(() => {
        T === 0 && mt(() => {
          o("nocLayoutReady", n.layout);
        }), n.responsive && A(), $();
      });
    }), V(() => n.useObserver, (w) => {
      if (!w) {
        D.disconnect();
        return;
      }
      U();
    });
    function y(w) {
      const T = {
        observe: [],
        unobserve: []
      };
      w.forEach(({ isIntersecting: z, target: S }) => {
        if (z) {
          T.observe.push(S[Ve]);
          return;
        }
        T.unobserve.push(S[Ve]);
      }), o("nocIntersectionObserve", T.observe), o("nocIntersectionUnobserve", T.unobserve);
    }
    function p(w) {
      const T = Object.keys(r);
      return Object.keys(w).reduce((z, S) => ((i.includes(S) || T.includes(S)) && (z[S] = w[S]), z), {});
    }
    function M() {
      if (n.layout && h.value) {
        if (n.layout.length !== h.value.length) {
          const w = x(n.layout, h.value);
          w.length > 0 && (n.layout.length > h.value.length ? h.value = h.value.concat(w) : h.value = h.value.filter((T) => !w.some((z) => T.id === z.id))), c.value = n.layout.length, F();
        }
        Dt(n.layout, n.verticalCompact), $(), o("nocLayoutUpdate", n.layout), _.value = !_.value;
      }
    }
    function x(w, T) {
      const z = w.filter((G) => !T.some((q) => G.id === q.id)), S = T.filter((G) => !w.some((q) => G.id === q.id));
      return z.concat(S);
    }
    function F() {
      u.value = Object.assign({}, n.responsiveLayouts);
    }
    function $() {
      const w = k();
      f.value = { height: w };
    }
    function k() {
      if (!n.autoSize || !n.layout)
        return;
      const [, w] = Yt(n.margin);
      return `${Ma(n.layout) * (n.rowHeight + w) + w}px`;
    }
    function W() {
      I.value && (v.value = I.value.offsetWidth);
    }
    function A() {
      const w = vc(n.breakpoints, v.value), T = qn(w, n.cols);
      l.value && !u.value[l.value] && (u.value[l.value] = Ke(n.layout));
      const z = mc(
        h.value,
        u.value,
        n.breakpoints,
        w,
        l.value,
        T,
        n.verticalCompact
      );
      u.value[w] = z, l.value !== w && o("update:breakpoints", w), l.value = w, o("update:layout", z), R.value = qn(w, n.cols);
    }
    function H() {
      o("nocLayoutCreate", n.layout);
    }
    function U() {
      D = new IntersectionObserver(y, {
        root: null,
        rootMargin: "8px",
        threshold: 0.4,
        ...n.intersectionObserverConfig
      });
    }
    H(), Kn(() => {
      gc("resize", W), s.value && I.value && s.value.uninstall(I.value);
    }), ai(() => {
      o("nocLayoutBeforeMount", n.layout);
    }), Jn(() => {
      o("nocLayoutMount", n.layout), mt(() => {
        h.value = n.layout, mt(() => {
          W(), F(), hc("resize", W.bind(this)), Dt(n.layout, n.verticalCompact), o("nocLayoutUpdate", n.layout), $(), I.value && s.value.listenTo(I.value, W), n.useObserver && U();
        });
      });
    });
    function b({ callback: w, eventType: T, h: z, id: S, w: G, x: q, y: B }) {
      const K = Fn(n.layout, S) ?? { ...r };
      T === "dragmove" || T === "dragstart" ? (m.value.id = S, m.value.x = K.x, m.value.y = K.y, m.value.w = G, m.value.h = z, mt(() => {
        a.value = !0;
      })) : mt(() => {
        a.value = !1;
      }), o("update:layout", Je(n.layout, K, q, B, !0, n.horizontalShift, n.preventCollision)), Dt(n.layout, n.verticalCompact), w(), $(), T === "dragend" && (Dt(n.layout, n.verticalCompact), o("nocLayoutUpdate", n.layout));
    }
    function L({ callback: w, eventType: T, h: z, id: S, w: G, x: q, y: B }) {
      const K = Fn(n.layout, S) ?? { ...r };
      let Ft;
      if (n.preventCollision) {
        const oe = Bo(n.layout, { ...K, h: z, w: G }).filter(
          (ct) => ct.id !== K.id
        );
        if (Ft = oe.length > 0, Ft) {
          let ct = Number.POSITIVE_INFINITY, E = Number.POSITIVE_INFINITY;
          oe.forEach((N) => {
            N.x > K.x && (ct = Math.min(ct, N.x)), N.y > K.y && (E = Math.min(E, N.y));
          }), Number.isFinite(ct) && (K.w = ct - K.x), Number.isFinite(E) && (K.h = E - K.y);
        }
      }
      Ft || (K.w = G, K.h = z), T === "resizestart" || T === "resizemove" ? (m.value.id = +S, m.value.x = q, m.value.y = B, m.value.w = K.w, m.value.h = K.h, mt(() => {
        a.value = !0;
      })) : mt(() => {
        a.value = !1;
      }), Dt(n.layout, n.verticalCompact), w(), $(), T === "resizeend" && o("nocLayoutUpdate", n.layout);
    }
    return ci(Ho, {
      handleDragEvent: b,
      handleResizeEvent: L
    }), (w, T) => (Gt(), de("div", null, [
      Zn("div", {
        ref_key: "wrapper",
        ref: I,
        class: "vue-grid-layout",
        style: Qn(f.value)
      }, [
        li(ui(Qe, gn({ class: "vue-grid-placeholder" }, { ...C.value, ...m.value }), null, 16), [
          [fi, a.value]
        ]),
        Ce(w.$slots, "default", {
          gridItemProps: { ...C.value, observer: Me(D) }
        }, () => [
          (Gt(!0), de(di, null, pi(w.layout, (z) => (Gt(), hi(Qe, gn({
            key: z.id
          }, { ...C.value, ...p(z) }, {
            observer: Me(D),
            onNocResizeContainer: T[0] || (T[0] = (S) => o("nocResizeContainer", S)),
            onNocResizeEnd: T[1] || (T[1] = (S) => o("nocItemResizeEnd", S)),
            onNocResize: T[2] || (T[2] = (S) => o("nocItemResize", S)),
            onNocMove: T[3] || (T[3] = (S) => o("nocItemMove", S)),
            onNocMoveEnd: T[4] || (T[4] = (S) => o("nocItemMoveEnd", S))
          }), {
            default: gi(() => [
              Ce(w.$slots, "gridItemContent", { item: z })
            ]),
            _: 2
          }, 1040, ["observer"]))), 128))
        ])
      ], 4)
    ]));
  }
});
function bc(e) {
  e.component("GridLayout", yc), e.component("GridItem", Qe);
}
const Sc = {
  install: bc
};
export {
  Qe as GridItem,
  yc as GridLayout,
  _t as MovingDirections,
  Sc as default
};
